---
title: "Explore HTAN Clinicial, Biospecimen, and Assay Metadata"
author: "Vesteinn Þórsson, ISB"
date: "Edited March 7, 2024"
output:
  html_document:
    keep_md: true
    highlight: tango
    toc: true
    toc_float:
      collapsed: TRUE
      smooth_scroll: false
---
```{r global options, echo = FALSE}
#List other global opts here if relevant
knitr::opts_chunk$set(fig.align="center") #Figures should be centered
## Utility to parse participant ID from other HTAN ID
participant_from_other_htan_id<- function(a){str_c(str_split(a,"_")[[1]][1:2],collapse="_")}
```

# 1. Introduction & Overview 

[HTAN](https://humantumoratlas.org/) is a National Cancer Institute (NCI)-funded Cancer Moonshot$^{SM}$ initiative to construct 3-dimensional atlases of the dynamic cellular, morphological, and molecular features of human cancers as they evolve from precancerous lesions to advanced disease ( [Cell April 2020](https://www.sciencedirect.com/science/article/pii/S0092867420303469) ).

Clinical data, sample biospecimen data and assay files in HTAN have a rich set of annotations supplied by HTAN data contributors.  These annotations are made according to the [HTAN Data model](https://data.humantumoratlas.org/standards), a set of standards defined by the HTAN consortium. The supplied values of these attributes have been collected into comprehensive data tables on the cloud, using the Google BigQuery structure that is part of Google Cloud Project.

### 1.1 Goal

This example notebook illustrates how to make use of HTAN Google BigQuery metadata tables to tabulate and plot available HTAN clinical, biospecimen, and assay metadata describing files available from [HTAN](https://data.humantumoratlas.org/). Summaries for other available metadata attributes can be generated by extending these examples.

### 1.2 Inputs, Outputs, & Data

The originating data can be found on the [HTAN Data Portal](https://data.humantumoratlas.org/), and the compiled tables are on the [Cancer Gateway in the Cloud](https://isb-cgc.appspot.com/).

### 1.3 Notes

The queries and results in this notebook correspond to HTAN Data Release 4. To choose a different release, edit the BigQuery table names in this notebook by replacing the string `r4` with a selected numbered release, e.g. `r2`. To get results for the most current data release, replace `r4` with `current` and `HTAN_versioned` with `HTAN`.  (For example replace `isb-cgc-bq.HTAN_versioned.clinical_tier1_demographics_r4` with `isb-cgc-bq.HTAN.clinical_tier1_demographics_current`).

# 2. Environment & Library Setup
```{r}
suppressMessages(library(tidyverse))
suppressMessages(library(bigrquery))
suppressMessages(library(knitr))
suppressMessages(library(reshape2))
```

# 3. Google Authentication

Running the BigQuery cells in this notebook requires a Google Cloud Project. Instructions for creating a project can be
found in [Google cloud documentation](https://cloud.google.com/resource-manager/docs/creating-managing-projects#console).
The instance needs to be authorized to bill the project for queries. For more information on getting started see
[Quick Start Guide to ISB-CGC](https://isb-cancer-genomics-cloud.readthedocs.io/en/latest/sections/HowToGetStartedonISB-CGC.html).
Alternative authentication methods can be found in the [Google Documentation](https://cloud.google.com/resource-manager/docs/creating-managing-projects#console).

```{r}

billing <- 'your_project_id' # Insert your project ID in the ''
if (billing == 'your_project_id') {
  print('Please update the project id with your Google Cloud Project')
}
```


# 4. Analyzing Clinical Data in HTAN

In the [HTAN Data model](https://data.humantumoratlas.org/standards), [Tier 1 Clinical Data](https://data.humantumoratlas.org/standard/clinical) has seven components: Demographics, Diagnosis, Exposure, Family History, Follow Up, Molecular Test, and Therapy. All HTAN demographic data is collected into a single Demographics table (*isb-cgc-bq.HTAN_versioned.clinical_tier1_demographics_r4*) in Google BigQuery containing data across all [HTAN Centers](https://humantumoratlas.org/research-network). The same is true of Diagnosis, and so on.

### 4.1 Demographics

Let's look at demographic distributions in HTAN. We begin by constructing an SQL query (as a string), then sending that as a query to HTAN Google BigQuery to retrieve the Demographics table. We remove a few unneeded bookkeeping columns.

```{r cache=TRUE}
sql  <- "select * from `isb-cgc-bq.HTAN_versioned.clinical_tier1_demographics_r4`"
tb <- bq_project_query(billing, sql)
demographics <- bq_table_download(tb)
demographics <- demographics %>% select(-entityId,-Component) %>% distinct()
demographics$HTAN_Center <- gsub("HTAN ","",demographics$HTAN_Center)
```

The number of rows of this table is the number of participants for which demographics is reported:
nrow(demographics)=`r nrow(demographics)`. The list of IDs of all HTAN Participants with demographic annotations is

```{r}
participant_list <- demographics$HTAN_Participant_ID 
```

There are length(participant_list)=`r length(participant_list)` participants with demographic annotations.

Participants per HTAN center
```{r echo=T}
participant_count_by_center <- demographics %>%
        select(HTAN_Center) %>%
        group_by(HTAN_Center) %>%
        tally()
participant_count_by_center %>% kable() 
```


You will see some differences between this table and the case counts on the [HTAN Data Portal]([HTAN](https://data.humantumoratlas.org/)).
(Some contributing factors: This notebook is fixed to HTAN data Release 2.0 while the portal has additional data;
the portal case inclusion criterion does not correspond to this simple row count.)

The number of columns in the Demographics table, ncol(demographics)=`r ncol(demographics)`, is the number of demographic
attributes. The attributes are
```{r}
colnames(demographics)
```
These are the attributes of the HTAN [Clinical Tier 1 Demographics Data Model](https://data.humantumoratlas.org/standard/clinical),
along with a few general attributes.

#### 4.1.1 Race
Race is one of the Demographics attributes. Let's tabulate reported Race in HTAN and report the fraction as a percent. 
```{r}
demographics_race_reported <- demographics %>%
        filter(Race != "unknown" & Race != "Unknown" & Race != "Not Reported" & Race != "not allowed to collect" )
demographics_race_reported %>%
        group_by(Race) %>%
        tally() %>%
        mutate(Percent=round(100*n/nrow(demographics_race_reported),1)) %>%
        kable()
```

Here is a barchart showing the distribution
```{r fig.dim = c(6, 3)}
demographics_race_reported %>% ggplot(aes(y=Race,fill=HTAN_Center)) + 
  geom_bar() + theme_classic() 
```

#### 4.1.2 Gender

Similarly, here's a barchart showing the breakdown by Gender.

```{r}
demographics_gender_reported <- demographics %>% filter(Race != "Not Reported") 
demographics_gender_reported %>%
        group_by(Gender) %>%
        tally() %>%
        mutate(Percent=round(100*n/nrow(demographics),1)) %>%
        kable()
```

```{r fig.dim = c(6, 3)}
demographics_gender_reported %>%
  ggplot(aes(y=Gender,fill=HTAN_Center)) +
  geom_bar() +
  theme_classic()
```

### 4.2 Treatment 

Let's take a look at annotated treatment in HTAN. This can be found in the Therapy table.

```{r cache=TRUE}
sql  <- "select * from `isb-cgc-bq.HTAN_versioned.clinical_tier1_therapy_r4`"
tb <- bq_project_query(billing, sql)
therapy <- bq_table_download(tb)
therapy <- therapy %>% select(-entityId,-Component) %>% distinct()
therapy$HTAN_Center <- gsub("HTAN ","",therapy$HTAN_Center)
```

Now filter this table to retrieve instances of annotated therapy. 

```{r}
therapy_yes <- therapy %>%
    filter((Treatment_or_Therapy == "Yes" | !is.na(.$Treatment_Type)) & Treatment_Type != "Not Reported" & Treatment_Type != "None")
```

These are `r nrow(therapy_yes)` in number, for `r length(unique(therapy_yes$HTAN_Participant_ID))` participants.


By center and treatment type:
```{r}
therapy_yes %>%
  group_by(HTAN_Center,Treatment_Type) %>%
  tally() %>% arrange(desc(n)) %>%
  kable()
```


# 5. Analyzing Biospecimen Data in HTAN

Biospecimen data in HTAN conforms to the [Biospecimen Data](https://data.humantumoratlas.org/standard/biospecimen) standard,
and all annotated values can be obtained in a single BigQuery table.

```{r, cache=TRUE}
sql <- "SELECT * FROM `isb-cgc-bq.HTAN_versioned.biospecimen_r4`"
tb <- bq_project_query(billing, sql)
biospecimen <- bq_table_download(tb)
biospecimen <- biospecimen %>% select(-entityId,-Component) 
biospecimen <- mutate(biospecimen,HTAN_Participant_ID=unlist(map(HTAN_Biospecimen_ID,participant_from_other_htan_id)))
```
Number of unique biospecimens : `r nrow(biospecimen)`, from `r length(unique(biospecimen$HTAN_Participant_ID))` participants.
Of these,  `r length(which(biospecimen$HTAN_Parent_ID %in% unique(demographics$HTAN_Participant_ID)))` list an HTAN Participant ID as Parent ID.

As an example, let's take a look at the various storage methods used for HTAN biospecimens

### 5.1 Storage Methods
```{r}
biospecimen_storage_noted <- biospecimen %>% filter(Storage_Method != "unknown" & Storage_Method != "Unknown")
table(biospecimen_storage_noted$Storage_Method,biospecimen_storage_noted$HTAN_Center) %>% kable()
```


# 6. Assay types in HTAN 

HTAN data is generated from multiple assay types, probing cancers and the tumor microenvironment at molecular, cellular, and tissue level,
using bulk, single-cell, and spatial assays. As described on the [HTAN Data Standards](https://data.humantumoratlas.org/standards)
page, generated data are arranged into data "Levels" corresponding to bioinformatic processing steps. In BigQuery, each
assay type and level is collected into a single table. 

For example, we can take a look at Single Cell RNA-seq Level 4 metadata by querying the table `isb-cgc-bq.HTAN_versioned.scRNAseq_level4_metadata_r4`

```{r, cache=TRUE, echo=T}
sql <- "SELECT * FROM `isb-cgc-bq.HTAN_versioned.scRNAseq_level4_metadata_r4`"
tb <- bq_project_query(billing, sql)
scl4 <- bq_table_download(tb)
scl4 %>% head(10) %>% kable() 
```

The [HTAN ID Provenance](https://isb-cgc.appspot.com/bq_meta_search/isb-cgc-bq.HTAN.id_provenance_current/) table pulls a selection of attributes from each individual assay metadata table, combining all of them into a single queryable table. For an in-depth introduction to this table, please see our notebook: [HTAN_ID_Provenance_In_BQ.ipynb](https://github.com/isb-cgc/Community-Notebooks/blob/master/HTAN/Python%20Notebooks/HTAN_ID_Provenance_In_BQ.ipynb).

Using the provenance table, we can tally the number of files available for each assay and level.

```{r, cache=TRUE, echo=T}
sql <- "SELECT Component, HTAN_Center, Count(*) AS Count
  FROM `isb-cgc-bq.HTAN_versioned.id_provenance_r4`
  WHERE Component LIKE '%Level%'
  GROUP BY Component, HTAN_Center
  ORDER BY Component, HTAN_Center"
tb <- bq_project_query(billing, sql)
all_files <- bq_table_download(tb)
```
```{r}
all_files_pivot <- dcast(all_files, Component ~ HTAN_Center, value.var = "Count", fun.aggregate = sum, fill = 0)
all_files_pivot %>% kable()
```


# 7. Relevant Citations and Links 

[Cell April 2020](https://www.sciencedirect.com/science/article/pii/S0092867420303469)


