{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "How_to_use_the_Reactome_BQ_dataset.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/isb-cgc/Community-Notebooks/blob/master/Notebooks/How_to_use_the_Reactome_BQ_dataset.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qgwv4lK5V8Zc"
      },
      "source": [
        "# How to use the Reactome BigQuery dataset\n",
        "Check out other notebooks at our [Community Notebooks Repository](https://github.com/isb-cgc/Community-Notebooks)!\n",
        "\n",
        "- **Title:** How to use the Reactome BigQuery dataset\n",
        "- **Author:** John Phan\n",
        "- **Created:** 2021-09-13\n",
        "- **Purpose:** Demonstrate basic usage of the Reactome BigQuery dataset\n",
        "- **URL:** https://github.com/isb-cgc/Community-Notebooks/blob/master/Notebooks/How_to_use_the_Reactome_BQ_dataset.ipynb\n",
        "\n",
        "This notebook demonstrates basic usage of the Reactome BigQuery dataset. Analysis of this dataset can provide a powerful tool for identifying pathways related to cancer biomarkers. \n",
        "\n",
        "The Reactome is an open-source, manually curated, and peer-reviewed pathway database. More information can be found here: https://reactome.org/. \n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1zVfEL8i8__O"
      },
      "source": [
        "# Initialize Notebook Environment\n",
        "\n",
        "Before running the analysis, we need to load dependencies, authenticate to BigQuery, and customize notebook parameters."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ujUW1wF23d93"
      },
      "source": [
        "## Import Dependencies"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LNlADvO-gken"
      },
      "source": [
        "# GCP Libraries\n",
        "from google.cloud import bigquery\n",
        "from google.colab import auth\n",
        "\n",
        "# Data Analytics\n",
        "import numpy as np\n",
        "from scipy import stats"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LBHnu8bZDEcE"
      },
      "source": [
        "## Authenticate\n",
        "\n",
        "Before using BigQuery, we need to get authorization for access to BigQuery and the Google Cloud. For more information see ['Quick Start Guide to ISB-CGC'](https://isb-cancer-genomics-cloud.readthedocs.io/en/latest/sections/HowToGetStartedonISB-CGC.html). Alternative authentication methods can be found [here](https://googleapis.dev/python/google-api-core/latest/auth.html)."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "M6BEFHj2dd5a"
      },
      "source": [
        "# if you're using Google Colab, authenticate to gcloud with the following\n",
        "auth.authenticate_user()\n",
        "\n",
        "# alternatively, use the gcloud SDK\n",
        "#!gcloud auth application-default login"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RWuECMsRptWJ"
      },
      "source": [
        "## Parameters\n",
        "\n",
        "Customize the following parameters based on your notebook, execution environment, or project."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "veP6gPatu2iW"
      },
      "source": [
        "# set the google project that will be billed for this notebook's computations\n",
        "google_project = 'google-project' ## change me"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "F6zTODbDq3g9"
      },
      "source": [
        "## BigQuery Client\n",
        "\n",
        "Create the BigQuery client."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HNkfSXedwzbC"
      },
      "source": [
        "# Create a client to access the data within BigQuery\n",
        "client = bigquery.Client(google_project)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7kVZ88dU85Vs"
      },
      "source": [
        "# Identify all Reactome Pathways Related to Genes\n",
        "\n",
        "We can join tables from the Reactome BigQuery dataset to identify all pathways related to our genes of interest. We first have to map the gene names to Uniprot IDs in the Reactome \"physical entities\" table. These can then be mapped to Reactome pathways. We further filter the physical entity to pathway evidence codes to retain only interactions that have \"Traceable Author Statements\" (TAS) rather than just \"Inferred from Electronic Annotation\" (IEA) to avoid evidence that have not been manually curated. \n",
        "\n",
        "We use the following genes to identify related pathways. These genes were identified in an ovarian cancer chemo-response study by [Bosquet et al](https://molecular-cancer.biomedcentral.com/articles/10.1186/s12943-016-0548-9).  "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4QDl8QK9dZJf"
      },
      "source": [
        "# set parameters for query\n",
        "genes = \"'RHOT1','MYO7A','ZBTB10','MATK','ST18','RPS23','GCNT1','DROSHA','NUAK1','CCPG1',\\\n",
        "'PDGFD','KLRAP1','MTAP','RNF13','THBS1','MLX','FAP','TIMP3','PRSS1','SLC7A11',\\\n",
        "'OLFML3','RPS20','MCM5','POLE','STEAP4','LRRC8D','WBP1L','ENTPD5','SYNE1','DPT',\\\n",
        "'COPZ2','TRIO','PDPR'\""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZwJGDXql9j6P"
      },
      "source": [
        "# run query and put results in data frame\n",
        "pathways = client.query((\"\"\"\n",
        "  SELECT\n",
        "    DISTINCT pathway.*\n",
        "\n",
        "  FROM\n",
        "    `isb-cgc-bq.reactome_versioned.pathway_v77` as pathway\n",
        "\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.pe_to_pathway_v77` as pe2pathway\n",
        "    -- link pathways to physical entities via intermediate table\n",
        "    ON pathway.stable_id = pe2pathway.pathway_stable_id\n",
        "\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
        "    -- link pathways to physical entities\n",
        "    ON pe2pathway.pe_stable_id = pe.stable_id\n",
        "\n",
        "  WHERE\n",
        "    -- filter by stronger evidence: \"Traceable Author Statement\" \n",
        "    pe2pathway.evidence_code = 'TAS'\n",
        "\n",
        "    -- filter by pathways that are related to genes in list\n",
        "    AND pe.name IN ({genes}) \n",
        "\n",
        "  ORDER BY pathway.name ASC\n",
        "\"\"\").format(\n",
        "  genes=genes\n",
        ")).result().to_dataframe()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 419
        },
        "id": "DmkVEYCHfcWK",
        "outputId": "dc724070-e5f7-40f5-fce8-b3bba4aec5a4"
      },
      "source": [
        "# Display the data frame\n",
        "pathways"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>stable_id</th>\n",
              "      <th>url</th>\n",
              "      <th>name</th>\n",
              "      <th>species</th>\n",
              "      <th>lowest_level</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>R-HSA-176187</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-17...</td>\n",
              "      <td>Activation of ATR in response to replication s...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>R-HSA-72662</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-72662</td>\n",
              "      <td>Activation of the mRNA upon binding of the cap...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>R-HSA-68962</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-68962</td>\n",
              "      <td>Activation of the pre-replicative complex</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>R-HSA-352230</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-35...</td>\n",
              "      <td>Amino acid transport across the plasma membrane</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>R-HSA-446203</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-44...</td>\n",
              "      <td>Asparagine N-linked glycosylation</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>148</th>\n",
              "      <td>R-HSA-192823</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-19...</td>\n",
              "      <td>Viral mRNA Translation</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>149</th>\n",
              "      <td>R-HSA-2187338</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-21...</td>\n",
              "      <td>Visual phototransduction</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>150</th>\n",
              "      <td>R-HSA-193704</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-19...</td>\n",
              "      <td>p75 NTR receptor-mediated signalling</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>151</th>\n",
              "      <td>R-HSA-72312</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-72312</td>\n",
              "      <td>rRNA processing</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>152</th>\n",
              "      <td>R-HSA-8868773</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-88...</td>\n",
              "      <td>rRNA processing in the nucleus and cytosol</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>153 rows × 5 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "         stable_id  ... lowest_level\n",
              "0     R-HSA-176187  ...         True\n",
              "1      R-HSA-72662  ...        False\n",
              "2      R-HSA-68962  ...         True\n",
              "3     R-HSA-352230  ...         True\n",
              "4     R-HSA-446203  ...        False\n",
              "..             ...  ...          ...\n",
              "148   R-HSA-192823  ...         True\n",
              "149  R-HSA-2187338  ...        False\n",
              "150   R-HSA-193704  ...        False\n",
              "151    R-HSA-72312  ...        False\n",
              "152  R-HSA-8868773  ...        False\n",
              "\n",
              "[153 rows x 5 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4IT_WO6PO0bH"
      },
      "source": [
        "If we're only interested in the lowest level pathways, i.e., pathways that are not parents of other pathways in the hierarchy, we can filter by the \"lowest_level\" field in the pathways table."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IQcGozM6OzpW"
      },
      "source": [
        "# run query and put results in data frame\n",
        "lowest_pathways = client.query((\"\"\"\n",
        "  SELECT\n",
        "    DISTINCT pathway.*\n",
        "\n",
        "  FROM\n",
        "    `isb-cgc-bq.reactome_versioned.pathway_v77` as pathway\n",
        "\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.pe_to_pathway_v77` as pe2pathway\n",
        "    -- link pathways to physical entities via intermediate table\n",
        "    ON pathway.stable_id = pe2pathway.pathway_stable_id\n",
        "\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
        "    -- link pathways to physical entities\n",
        "    ON pe2pathway.pe_stable_id = pe.stable_id\n",
        "\n",
        "  WHERE\n",
        "    -- filter by stronger evidence: \"Traceable Author Statement\" \n",
        "    pe2pathway.evidence_code = 'TAS'\n",
        "\n",
        "    -- filter by pathways that are related to genes in list\n",
        "    AND pe.name IN ({genes})\n",
        "\n",
        "    -- filter to include just lowest level pathways\n",
        "    AND pathway.lowest_level = TRUE\n",
        "\n",
        "  ORDER BY pathway.name ASC\n",
        "\"\"\").format(\n",
        "  genes=genes\n",
        ")).result().to_dataframe()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "q5bZg8w6Pn7Q",
        "outputId": "8096e367-7452-4814-d783-90df5cda9632"
      },
      "source": [
        "# display data frame\n",
        "lowest_pathways"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>stable_id</th>\n",
              "      <th>url</th>\n",
              "      <th>name</th>\n",
              "      <th>species</th>\n",
              "      <th>lowest_level</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>R-HSA-176187</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-17...</td>\n",
              "      <td>Activation of ATR in response to replication s...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>R-HSA-68962</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-68962</td>\n",
              "      <td>Activation of the pre-replicative complex</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>R-HSA-352230</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-35...</td>\n",
              "      <td>Amino acid transport across the plasma membrane</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>R-HSA-210991</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-21...</td>\n",
              "      <td>Basigin interactions</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>R-HSA-9013148</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-90...</td>\n",
              "      <td>CDC42 GTPase cycle</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>R-HSA-6811434</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-68...</td>\n",
              "      <td>COPI-dependent Golgi-to-ER retrograde traffic</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>R-HSA-6807878</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-68...</td>\n",
              "      <td>COPI-mediated anterograde transport</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>R-HSA-163765</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-16...</td>\n",
              "      <td>ChREBP activates metabolic gene expression</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>R-HSA-418885</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-41...</td>\n",
              "      <td>DCC mediated attractive signaling</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>R-HSA-68952</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-68952</td>\n",
              "      <td>DNA replication initiation</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>R-HSA-5696400</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>Dual Incision in GG-NER</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>R-HSA-6782135</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-67...</td>\n",
              "      <td>Dual incision in TC-NER</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>R-HSA-72689</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-72689</td>\n",
              "      <td>Formation of a pool of free 40S subunits</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>R-HSA-72695</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-72695</td>\n",
              "      <td>Formation of the ternary complex, and subseque...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>R-HSA-416482</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-41...</td>\n",
              "      <td>G alpha (12/13) signalling events</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>R-HSA-72706</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-72706</td>\n",
              "      <td>GTP hydrolysis and joining of the 60S ribosoma...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>R-HSA-5696397</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>R-HSA-6782210</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-67...</td>\n",
              "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>R-HSA-8950505</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-89...</td>\n",
              "      <td>Gene and protein expression by JAK-STAT signal...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>R-HSA-216083</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-21...</td>\n",
              "      <td>Integrin cell surface interactions</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>R-HSA-156827</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-15...</td>\n",
              "      <td>L13a-mediated translational silencing of Cerul...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>R-HSA-6791226</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-67...</td>\n",
              "      <td>Major pathway of rRNA processing in the nucleo...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>R-HSA-1237112</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-12...</td>\n",
              "      <td>Methionine salvage pathway</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>R-HSA-203927</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-20...</td>\n",
              "      <td>MicroRNA (miRNA) biogenesis</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>R-HSA-5223345</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-52...</td>\n",
              "      <td>Miscellaneous transport and binding events</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>R-HSA-193648</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-19...</td>\n",
              "      <td>NRAGE signals death through JNK</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>R-HSA-975957</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-97...</td>\n",
              "      <td>Nonsense Mediated Decay (NMD) enhanced by the ...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>R-HSA-975956</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-97...</td>\n",
              "      <td>Nonsense Mediated Decay (NMD) independent of t...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>R-HSA-5173214</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-51...</td>\n",
              "      <td>O-glycosylation of TSR domain-containing proteins</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>R-HSA-68949</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-68949</td>\n",
              "      <td>Orc1 removal from chromatin</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>R-HSA-5651801</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>PCNA-Dependent Long Patch Base Excision Repair</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>R-HSA-8850843</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-88...</td>\n",
              "      <td>Phosphate bond hydrolysis by NTPDase proteins</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>R-HSA-114608</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-11...</td>\n",
              "      <td>Platelet degranulation</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>R-HSA-9660826</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-96...</td>\n",
              "      <td>Purinergic signaling in leishmaniasis infection</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>R-HSA-9013149</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-90...</td>\n",
              "      <td>RAC1 GTPase cycle</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>R-HSA-9013404</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-90...</td>\n",
              "      <td>RAC2 GTPase cycle</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>R-HSA-9013423</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-90...</td>\n",
              "      <td>RAC3 GTPase cycle</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>R-HSA-8980692</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-89...</td>\n",
              "      <td>RHOA GTPase cycle</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>R-HSA-9013408</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-90...</td>\n",
              "      <td>RHOG GTPase cycle</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>R-HSA-9013409</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-90...</td>\n",
              "      <td>RHOJ GTPase cycle</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>40</th>\n",
              "      <td>R-HSA-9013425</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-90...</td>\n",
              "      <td>RHOT1 GTPase cycle</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>R-HSA-8936459</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-89...</td>\n",
              "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>R-HSA-110314</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-11...</td>\n",
              "      <td>Recognition of DNA damage by PCNA-containing r...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>R-HSA-6804756</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-68...</td>\n",
              "      <td>Regulation of TP53 Activity through Phosphoryl...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>44</th>\n",
              "      <td>R-HSA-204174</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-20...</td>\n",
              "      <td>Regulation of pyruvate dehydrogenase (PDH) com...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>R-HSA-72702</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-72702</td>\n",
              "      <td>Ribosomal scanning and start codon recognition</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>R-HSA-1799339</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-17...</td>\n",
              "      <td>SRP-dependent cotranslational protein targetin...</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47</th>\n",
              "      <td>R-HSA-5656169</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>Termination of translesion DNA synthesis</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>R-HSA-72649</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-72649</td>\n",
              "      <td>Translation initiation complex formation</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>49</th>\n",
              "      <td>R-HSA-5689880</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>Ub-specific processing proteases</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50</th>\n",
              "      <td>R-HSA-192823</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-19...</td>\n",
              "      <td>Viral mRNA Translation</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "        stable_id  ... lowest_level\n",
              "0    R-HSA-176187  ...         True\n",
              "1     R-HSA-68962  ...         True\n",
              "2    R-HSA-352230  ...         True\n",
              "3    R-HSA-210991  ...         True\n",
              "4   R-HSA-9013148  ...         True\n",
              "5   R-HSA-6811434  ...         True\n",
              "6   R-HSA-6807878  ...         True\n",
              "7    R-HSA-163765  ...         True\n",
              "8    R-HSA-418885  ...         True\n",
              "9     R-HSA-68952  ...         True\n",
              "10  R-HSA-5696400  ...         True\n",
              "11  R-HSA-6782135  ...         True\n",
              "12    R-HSA-72689  ...         True\n",
              "13    R-HSA-72695  ...         True\n",
              "14   R-HSA-416482  ...         True\n",
              "15    R-HSA-72706  ...         True\n",
              "16  R-HSA-5696397  ...         True\n",
              "17  R-HSA-6782210  ...         True\n",
              "18  R-HSA-8950505  ...         True\n",
              "19   R-HSA-216083  ...         True\n",
              "20   R-HSA-156827  ...         True\n",
              "21  R-HSA-6791226  ...         True\n",
              "22  R-HSA-1237112  ...         True\n",
              "23   R-HSA-203927  ...         True\n",
              "24  R-HSA-5223345  ...         True\n",
              "25   R-HSA-193648  ...         True\n",
              "26   R-HSA-975957  ...         True\n",
              "27   R-HSA-975956  ...         True\n",
              "28  R-HSA-5173214  ...         True\n",
              "29    R-HSA-68949  ...         True\n",
              "30  R-HSA-5651801  ...         True\n",
              "31  R-HSA-8850843  ...         True\n",
              "32   R-HSA-114608  ...         True\n",
              "33  R-HSA-9660826  ...         True\n",
              "34  R-HSA-9013149  ...         True\n",
              "35  R-HSA-9013404  ...         True\n",
              "36  R-HSA-9013423  ...         True\n",
              "37  R-HSA-8980692  ...         True\n",
              "38  R-HSA-9013408  ...         True\n",
              "39  R-HSA-9013409  ...         True\n",
              "40  R-HSA-9013425  ...         True\n",
              "41  R-HSA-8936459  ...         True\n",
              "42   R-HSA-110314  ...         True\n",
              "43  R-HSA-6804756  ...         True\n",
              "44   R-HSA-204174  ...         True\n",
              "45    R-HSA-72702  ...         True\n",
              "46  R-HSA-1799339  ...         True\n",
              "47  R-HSA-5656169  ...         True\n",
              "48    R-HSA-72649  ...         True\n",
              "49  R-HSA-5689880  ...         True\n",
              "50   R-HSA-192823  ...         True\n",
              "\n",
              "[51 rows x 5 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ybqSP2aVNE99"
      },
      "source": [
        "# Pathway Enrichment Analysis\n",
        "We can identify pathways that are \"enriched\" with the genes of interest. In other words, we can answer the question: given a set of interesting genes, which pathways contain those genes at a frequency higher than random chance? By calculating the probability that a number of target genes are contained in each pathway, we can identify pathways most likely to be related. \n",
        "\n",
        "To do this, we can use a **chi-squared** test to determine if there is a statistically significant difference between the expected frequency of genes in a pathway compared to the observed frequency. \n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sKLy2sxQNIu6"
      },
      "source": [
        "# set query parameters\n",
        "lowest_level = True # only show pathways at the lowest level"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jDdXWbOHa32V"
      },
      "source": [
        "## Construct SQL Query\n",
        "\n",
        "A single query can be used to calculate the chi-squared statistic for all pathways. This query is rather lengthy, but can be broken up into a series of named sub-queries. We step through each query below.\n",
        "\n",
        "First, we write a query that simply gets a list of all genes in the Reactome physical entity table:  "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "r3NeGvfMbgBp"
      },
      "source": [
        "gene_list_query = \"\"\"\n",
        "  -- Table that contains a list of all distinct genes that map to Reactome\n",
        "  -- physical entities \n",
        "  SELECT\n",
        "    DISTINCT pe.uniprot_id\n",
        "\n",
        "  FROM\n",
        "    `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
        "\n",
        "  WHERE\n",
        "    -- filter by pathways that are related to genes in list\n",
        "    pe.name IN ({genes})\n",
        "\"\"\".format(\n",
        "  genes=genes\n",
        ").strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wTVNQFJjgQ1y",
        "outputId": "b089723a-6c24-4185-9c43-8e34d02f6a55"
      },
      "source": [
        "print(gene_list_query)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  -- Table that contains a list of all distinct genes that map to Reactome\n",
            "  -- physical entities \n",
            "  SELECT\n",
            "    DISTINCT pe.uniprot_id\n",
            "\n",
            "  FROM\n",
            "    `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
            "\n",
            "  WHERE\n",
            "    -- filter by pathways that are related to genes in list\n",
            "    pe.name IN ('RHOT1','MYO7A','ZBTB10','MATK','ST18','RPS23','GCNT1','DROSHA','NUAK1','CCPG1','PDGFD','KLRAP1','MTAP','RNF13','THBS1','MLX','FAP','TIMP3','PRSS1','SLC7A11','OLFML3','RPS20','MCM5','POLE','STEAP4','LRRC8D','WBP1L','ENTPD5','SYNE1','DPT','COPZ2','TRIO','PDPR')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bY4QCtDfgwBq"
      },
      "source": [
        "We then create a query that counts all targets associated with each Reactome pathway. This query depends on the previous `gene_list_query` sub-query.  "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "H_3xAQdjhJ9s"
      },
      "source": [
        "gene_pp_query = \"\"\"\n",
        "  -- Table that maps pathways to the total number of interesting genes within\n",
        "  -- that pathway\n",
        "  SELECT\n",
        "    COUNT(DISTINCT gene_list_query.uniprot_id) as num_genes,\n",
        "    pathway.stable_id,\n",
        "    pathway.name\n",
        "\n",
        "  FROM\n",
        "    gene_list_query\n",
        "\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
        "    -- filter for interactions with genes that match a reactome\n",
        "    -- physical entity\n",
        "    ON gene_list_query.uniprot_id = pe.uniprot_id\n",
        "\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.pe_to_pathway_v77` AS pe2pathway\n",
        "    -- link physical entities to pathways via intermediate table\n",
        "    ON pe.stable_id = pe2pathway.pe_stable_id\n",
        "\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.pathway_v77` AS pathway\n",
        "    -- link physical entities to pathways\n",
        "    ON pe2pathway.pathway_stable_id = pathway.stable_id\n",
        "\n",
        "  WHERE\n",
        "    -- filter by stronger evidence: \"Traceable Author Statement\" \n",
        "    pe2pathway.evidence_code = 'TAS'\n",
        "\n",
        "  GROUP BY pathway.stable_id, pathway.name\n",
        "  ORDER BY num_genes DESC\n",
        "\"\"\".strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "v29oHYEukrzy",
        "outputId": "975b1f58-969e-4676-d5b7-45ab0b78694c"
      },
      "source": [
        "print(gene_pp_query)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  -- Table that maps pathways to the total number of interesting genes within\n",
            "  -- that pathway\n",
            "  SELECT\n",
            "    COUNT(DISTINCT gene_list_query.uniprot_id) as num_genes,\n",
            "    pathway.stable_id,\n",
            "    pathway.name\n",
            "\n",
            "  FROM\n",
            "    gene_list_query\n",
            "\n",
            "  INNER JOIN `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
            "    -- filter for interactions with genes that match a reactome\n",
            "    -- physical entity\n",
            "    ON gene_list_query.uniprot_id = pe.uniprot_id\n",
            "\n",
            "  INNER JOIN `isb-cgc-bq.reactome_versioned.pe_to_pathway_v77` AS pe2pathway\n",
            "    -- link physical entities to pathways via intermediate table\n",
            "    ON pe.stable_id = pe2pathway.pe_stable_id\n",
            "\n",
            "  INNER JOIN `isb-cgc-bq.reactome_versioned.pathway_v77` AS pathway\n",
            "    -- link physical entities to pathways\n",
            "    ON pe2pathway.pathway_stable_id = pathway.stable_id\n",
            "\n",
            "  WHERE\n",
            "    -- filter by stronger evidence: \"Traceable Author Statement\" \n",
            "    pe2pathway.evidence_code = 'TAS'\n",
            "\n",
            "  GROUP BY pathway.stable_id, pathway.name\n",
            "  ORDER BY num_genes DESC\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_XXiAJZUlBxC"
      },
      "source": [
        "Now we construct the same queries for genes that are NOT in the interesting gene list. These are prefixed with \"not_gene\". This query depends on the previous `gene_list_query` sub-query. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1KyR-_WqlVVn"
      },
      "source": [
        "not_gene_list_query = \"\"\"\n",
        "  -- Table that contains a list of all genes that are NOT in the interest list\n",
        "  -- This query depends on the previous \"gene_list_query\" sub-query.\n",
        "  SELECT\n",
        "    DISTINCT pe.uniprot_id\n",
        "\n",
        "  FROM `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
        "\n",
        "  WHERE\n",
        "    pe.uniprot_id NOT IN (\n",
        "      SELECT uniprot_id FROM gene_list_query\n",
        "    )\n",
        "\"\"\".strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "92MaTzsAn3es",
        "outputId": "c50ae34a-1cbf-496e-86b4-ab804bc9c581"
      },
      "source": [
        "print(not_gene_list_query)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  -- Table that contains a list of all genes that are NOT in the interest list\n",
            "  -- This query depends on the previous \"gene_list_query\" sub-query.\n",
            "  SELECT\n",
            "    DISTINCT pe.uniprot_id\n",
            "\n",
            "  FROM `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
            "\n",
            "  WHERE\n",
            "    pe.uniprot_id NOT IN (\n",
            "      SELECT uniprot_id FROM gene_list_query\n",
            "    )\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "mI1SxkGtn-ow"
      },
      "source": [
        "not_gene_pp_query = \"\"\"\n",
        "  -- Table that maps pathways to the number of proteins that are NOT drug\n",
        "  -- targets in that pathway.\n",
        "  SELECT\n",
        "    COUNT(DISTINCT not_gene_list_query.uniprot_id) AS num_not_genes,\n",
        "    pathway.stable_id,\n",
        "    pathway.name\n",
        "\n",
        "  FROM not_gene_list_query\n",
        "\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
        "    ON not_gene_list_query.uniprot_id = pe.uniprot_id\n",
        "\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.pe_to_pathway_v77` AS pe2pathway\n",
        "    ON pe.stable_id = pe2pathway.pe_stable_id\n",
        "  \n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.pathway_v77` AS pathway\n",
        "    ON pe2pathway.pathway_stable_id = pathway.stable_id\n",
        "\n",
        "  WHERE\n",
        "    -- filter by stronger evidence: \"Traceable Author Statement\" \n",
        "    pe2pathway.evidence_code = 'TAS'\n",
        "\n",
        "  GROUP BY pathway.stable_id, pathway.name\n",
        "  ORDER BY num_not_genes DESC\n",
        "\"\"\".strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TgY33Z3ukByC",
        "outputId": "e9143dce-28de-4f43-ce7a-032c79757486"
      },
      "source": [
        "print(not_gene_pp_query)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  -- Table that maps pathways to the number of proteins that are NOT drug\n",
            "  -- targets in that pathway.\n",
            "  SELECT\n",
            "    COUNT(DISTINCT not_gene_list_query.uniprot_id) AS num_not_genes,\n",
            "    pathway.stable_id,\n",
            "    pathway.name\n",
            "\n",
            "  FROM not_gene_list_query\n",
            "\n",
            "  INNER JOIN `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
            "    ON not_gene_list_query.uniprot_id = pe.uniprot_id\n",
            "\n",
            "  INNER JOIN `isb-cgc-bq.reactome_versioned.pe_to_pathway_v77` AS pe2pathway\n",
            "    ON pe.stable_id = pe2pathway.pe_stable_id\n",
            "  \n",
            "  INNER JOIN `isb-cgc-bq.reactome_versioned.pathway_v77` AS pathway\n",
            "    ON pe2pathway.pathway_stable_id = pathway.stable_id\n",
            "\n",
            "  WHERE\n",
            "    -- filter by stronger evidence: \"Traceable Author Statement\" \n",
            "    pe2pathway.evidence_code = 'TAS'\n",
            "\n",
            "  GROUP BY pathway.stable_id, pathway.name\n",
            "  ORDER BY num_not_genes DESC\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Z2eN5_UfVgcX"
      },
      "source": [
        "For convenience, we create a sub-query to get the counts of the number of genes that are in the list and the number of genes that are not in the list."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "VAYDIHsdWLFA"
      },
      "source": [
        "gene_count_query = \"\"\"\n",
        "  -- Table that contains the counts of # of genes that are/are not targets\n",
        "  SELECT\n",
        "    gene_count,\n",
        "    not_gene_count,\n",
        "    gene_count + not_gene_count AS total_count\n",
        "\n",
        "  FROM \n",
        "    (SELECT COUNT(*) AS gene_count FROM gene_list_query),\n",
        "    (SELECT COUNT(*) AS not_gene_count FROM not_gene_list_query)\n",
        "\"\"\".strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qrt8DcUraJm3"
      },
      "source": [
        "Now we can create more interesting queries for the contingency matrices that contain the observed and expected values. \n",
        "\n",
        "First, the observed contingency table counts for each pathway:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "8Dc-wMSZakad"
      },
      "source": [
        "observed_query = \"\"\"\n",
        "  -- Table with observed values per pathway in the contingency matrix\n",
        "  SELECT\n",
        "    gene_pp_query.num_genes AS in_gene_in_pathway,\n",
        "    not_gene_pp_query.num_not_genes AS not_gene_in_pathway,\n",
        "    gene_count_query.gene_count - gene_pp_query.num_genes AS in_gene_not_pathway,\n",
        "    gene_count_query.not_gene_count - not_gene_pp_query.num_not_genes AS not_gene_not_pathway,\n",
        "    gene_pp_query.stable_id,\n",
        "    gene_pp_query.name\n",
        "\n",
        "  FROM \n",
        "    gene_pp_query,\n",
        "    gene_count_query\n",
        "\n",
        "  INNER JOIN not_gene_pp_query\n",
        "    ON gene_pp_query.stable_id = not_gene_pp_query.stable_id\n",
        "\"\"\".strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "STbl3Toxcq7B"
      },
      "source": [
        "Then the observed row and column sums of the contingency table:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4kXNQ4SRcmPK"
      },
      "source": [
        "sum_query = \"\"\"\n",
        "  -- Table with summed observed values per pathway in the contingency matrix\n",
        "  SELECT\n",
        "    observed_query.in_gene_in_pathway + observed_query.not_gene_in_pathway AS pathway_total,\n",
        "    observed_query.in_gene_not_pathway + observed_query.not_gene_not_pathway AS not_pathway_total,\n",
        "    observed_query.in_gene_in_pathway + observed_query.in_gene_not_pathway AS gene_total,\n",
        "    observed_query.not_gene_in_pathway + observed_query.not_gene_not_pathway AS not_gene_total,\n",
        "    observed_query.stable_id,\n",
        "    observed_query.name\n",
        "\n",
        "  FROM\n",
        "    observed_query\n",
        "\"\"\".strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6HiQlxAudQEk"
      },
      "source": [
        "And the expected contingency table values for each pathway:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "V0Rj098QdWe6"
      },
      "source": [
        "expected_query = \"\"\"  \n",
        "  -- Table with the expected values per pathway in the contingency matrix\n",
        "  SELECT \n",
        "    sum_query.gene_total * sum_query.pathway_total / gene_count_query.total_count AS exp_in_gene_in_pathway,\n",
        "    sum_query.not_gene_total * sum_query.pathway_total / gene_count_query.total_count AS exp_not_gene_in_pathway,\n",
        "    sum_query.gene_total * sum_query.not_pathway_total / gene_count_query.total_count AS exp_in_gene_not_pathway,\n",
        "    sum_query.not_gene_total * sum_query.not_pathway_total / gene_count_query.total_count AS exp_not_gene_not_pathway,\n",
        "    sum_query.stable_id,\n",
        "    sum_query.name\n",
        "\n",
        "  FROM \n",
        "    sum_query, gene_count_query\n",
        "\"\"\".strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wC57CDA6d_4H"
      },
      "source": [
        "Finally, we can calculate the chi-squared statistic for each pathway:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kxlrSyiTd_Ia"
      },
      "source": [
        "chi_squared_query = \"\"\"\n",
        "  -- Table with the chi-squared statistic for each pathway\n",
        "  SELECT\n",
        "    -- Chi squared statistic with Yates' correction\n",
        "    POW(ABS(observed_query.in_gene_in_pathway - expected_query.exp_in_gene_in_pathway) - 0.5, 2) / expected_query.exp_in_gene_in_pathway \n",
        "    + POW(ABS(observed_query.not_gene_in_pathway - expected_query.exp_not_gene_in_pathway) - 0.5, 2) / expected_query.exp_not_gene_in_pathway\n",
        "    + POW(ABS(observed_query.in_gene_not_pathway - expected_query.exp_in_gene_not_pathway) - 0.5, 2) / expected_query.exp_in_gene_not_pathway\n",
        "    + POW(ABS(observed_query.not_gene_not_pathway - expected_query.exp_not_gene_not_pathway) - 0.5, 2) / expected_query.exp_not_gene_not_pathway\n",
        "    AS chi_squared_stat,\n",
        "    observed_query.stable_id,\n",
        "    observed_query.name\n",
        "\n",
        "  FROM observed_query\n",
        "\n",
        "  INNER JOIN expected_query\n",
        "    ON observed_query.stable_id = expected_query.stable_id\n",
        "\"\"\".strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lTRfNbLKuarY"
      },
      "source": [
        "The final piece of the query optionally adds a filter that removes all pathways that are not at the lowest level of the hierarchy. This helps remove non-specific \"pathways\" such as the generic \"Disease\" pathway."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IBpHKWxjvCSx"
      },
      "source": [
        "lowest_level_filter = \"\"\"\n",
        "  INNER JOIN `isb-cgc-bq.reactome_versioned.pathway_v77` AS pathway\n",
        "    ON chi_squared_query.stable_id = pathway.stable_id\n",
        "\n",
        "  WHERE pathway.lowest_level = TRUE\n",
        "\"\"\".strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "V_Qx1J2IvrBT",
        "outputId": "04ee2a9a-a1df-416f-fef6-83cb0bc23671"
      },
      "source": [
        "print(lowest_level_filter)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  INNER JOIN `isb-cgc-bq.reactome_versioned.pathway_v77` AS pathway\n",
            "    ON chi_squared_query.stable_id = pathway.stable_id\n",
            "\n",
            "  WHERE pathway.lowest_level = TRUE\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QsJQyREafSZE"
      },
      "source": [
        "Now we can combine all sub-queries to create the final query:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dOXXXAmkfcnb"
      },
      "source": [
        "final_query = \"\"\"\n",
        "  WITH\n",
        "    gene_list_query AS (\n",
        "      {gene_list_query}\n",
        "    ),\n",
        "\n",
        "    gene_pp_query AS (\n",
        "      {gene_pp_query}\n",
        "    ),\n",
        "\n",
        "    not_gene_list_query AS (\n",
        "      {not_gene_list_query}\n",
        "    ),\n",
        "\n",
        "    not_gene_pp_query AS (\n",
        "      {not_gene_pp_query}\n",
        "    ),\n",
        "\n",
        "    gene_count_query AS (\n",
        "      {gene_count_query}\n",
        "    ),\n",
        "\n",
        "    observed_query AS (\n",
        "      {observed_query}\n",
        "    ),\n",
        "\n",
        "    sum_query AS (\n",
        "      {sum_query}\n",
        "    ),\n",
        "\n",
        "    expected_query AS (\n",
        "      {expected_query}\n",
        "    ),\n",
        "    \n",
        "    chi_squared_query AS (\n",
        "      {chi_squared_query}\n",
        "    )\n",
        "\n",
        "  SELECT\n",
        "    observed_query.in_gene_in_pathway,\n",
        "    observed_query.in_gene_not_pathway,\n",
        "    observed_query.not_gene_in_pathway,\n",
        "    observed_query.not_gene_not_pathway,\n",
        "    chi_squared_query.chi_squared_stat,\n",
        "    chi_squared_query.stable_id,\n",
        "    chi_squared_query.name\n",
        "\n",
        "  FROM chi_squared_query\n",
        "\n",
        "  INNER JOIN observed_query\n",
        "    ON chi_squared_query.stable_id = observed_query.stable_id\n",
        "  {lowest_level_filter}\n",
        "  ORDER BY chi_squared_stat DESC\n",
        "\"\"\".format(\n",
        "  # make final query a little easier to read by removing/adding some white space \n",
        "  gene_list_query=\"\\n    \".join(gene_list_query.strip().splitlines()),\n",
        "  gene_pp_query=\"\\n    \".join(gene_pp_query.strip().splitlines()),\n",
        "  not_gene_list_query=\"\\n    \".join(not_gene_list_query.strip().splitlines()),\n",
        "  not_gene_pp_query=\"\\n    \".join(not_gene_pp_query.strip().splitlines()),\n",
        "  gene_count_query=\"\\n    \".join(gene_count_query.strip().splitlines()),\n",
        "  observed_query=\"\\n    \".join(observed_query.strip().splitlines()),\n",
        "  sum_query=\"\\n    \".join(sum_query.strip().splitlines()),\n",
        "  expected_query=\"\\n    \".join(expected_query.strip().splitlines()),\n",
        "  chi_squared_query=\"\\n    \".join(chi_squared_query.strip().splitlines()),\n",
        "  lowest_level_filter=(\n",
        "    \"\\n  \"+\"\\n\".join(lowest_level_filter.strip().splitlines())+\"\\n\" if lowest_level else \"\"\n",
        "  )\n",
        ").strip(\"\\n\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YEjfpM9Q4xsz"
      },
      "source": [
        "## Display Final Query"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MfitLyzmgq_P",
        "outputId": "b554ac13-3c67-4639-9993-e1df53834ee5"
      },
      "source": [
        "# print the formatted final query\n",
        "print(final_query)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  WITH\n",
            "    gene_list_query AS (\n",
            "      -- Table that contains a list of all distinct genes that map to Reactome\n",
            "      -- physical entities \n",
            "      SELECT\n",
            "        DISTINCT pe.uniprot_id\n",
            "    \n",
            "      FROM\n",
            "        `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
            "    \n",
            "      WHERE\n",
            "        -- filter by pathways that are related to genes in list\n",
            "        pe.name IN ('RHOT1','MYO7A','ZBTB10','MATK','ST18','RPS23','GCNT1','DROSHA','NUAK1','CCPG1','PDGFD','KLRAP1','MTAP','RNF13','THBS1','MLX','FAP','TIMP3','PRSS1','SLC7A11','OLFML3','RPS20','MCM5','POLE','STEAP4','LRRC8D','WBP1L','ENTPD5','SYNE1','DPT','COPZ2','TRIO','PDPR')\n",
            "    ),\n",
            "\n",
            "    gene_pp_query AS (\n",
            "      -- Table that maps pathways to the total number of interesting genes within\n",
            "      -- that pathway\n",
            "      SELECT\n",
            "        COUNT(DISTINCT gene_list_query.uniprot_id) as num_genes,\n",
            "        pathway.stable_id,\n",
            "        pathway.name\n",
            "    \n",
            "      FROM\n",
            "        gene_list_query\n",
            "    \n",
            "      INNER JOIN `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
            "        -- filter for interactions with genes that match a reactome\n",
            "        -- physical entity\n",
            "        ON gene_list_query.uniprot_id = pe.uniprot_id\n",
            "    \n",
            "      INNER JOIN `isb-cgc-bq.reactome_versioned.pe_to_pathway_v77` AS pe2pathway\n",
            "        -- link physical entities to pathways via intermediate table\n",
            "        ON pe.stable_id = pe2pathway.pe_stable_id\n",
            "    \n",
            "      INNER JOIN `isb-cgc-bq.reactome_versioned.pathway_v77` AS pathway\n",
            "        -- link physical entities to pathways\n",
            "        ON pe2pathway.pathway_stable_id = pathway.stable_id\n",
            "    \n",
            "      WHERE\n",
            "        -- filter by stronger evidence: \"Traceable Author Statement\" \n",
            "        pe2pathway.evidence_code = 'TAS'\n",
            "    \n",
            "      GROUP BY pathway.stable_id, pathway.name\n",
            "      ORDER BY num_genes DESC\n",
            "    ),\n",
            "\n",
            "    not_gene_list_query AS (\n",
            "      -- Table that contains a list of all genes that are NOT in the interest list\n",
            "      -- This query depends on the previous \"gene_list_query\" sub-query.\n",
            "      SELECT\n",
            "        DISTINCT pe.uniprot_id\n",
            "    \n",
            "      FROM `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
            "    \n",
            "      WHERE\n",
            "        pe.uniprot_id NOT IN (\n",
            "          SELECT uniprot_id FROM gene_list_query\n",
            "        )\n",
            "    ),\n",
            "\n",
            "    not_gene_pp_query AS (\n",
            "      -- Table that maps pathways to the number of proteins that are NOT drug\n",
            "      -- targets in that pathway.\n",
            "      SELECT\n",
            "        COUNT(DISTINCT not_gene_list_query.uniprot_id) AS num_not_genes,\n",
            "        pathway.stable_id,\n",
            "        pathway.name\n",
            "    \n",
            "      FROM not_gene_list_query\n",
            "    \n",
            "      INNER JOIN `isb-cgc-bq.reactome_versioned.physical_entity_v77` AS pe\n",
            "        ON not_gene_list_query.uniprot_id = pe.uniprot_id\n",
            "    \n",
            "      INNER JOIN `isb-cgc-bq.reactome_versioned.pe_to_pathway_v77` AS pe2pathway\n",
            "        ON pe.stable_id = pe2pathway.pe_stable_id\n",
            "      \n",
            "      INNER JOIN `isb-cgc-bq.reactome_versioned.pathway_v77` AS pathway\n",
            "        ON pe2pathway.pathway_stable_id = pathway.stable_id\n",
            "    \n",
            "      WHERE\n",
            "        -- filter by stronger evidence: \"Traceable Author Statement\" \n",
            "        pe2pathway.evidence_code = 'TAS'\n",
            "    \n",
            "      GROUP BY pathway.stable_id, pathway.name\n",
            "      ORDER BY num_not_genes DESC\n",
            "    ),\n",
            "\n",
            "    gene_count_query AS (\n",
            "      -- Table that contains the counts of # of genes that are/are not targets\n",
            "      SELECT\n",
            "        gene_count,\n",
            "        not_gene_count,\n",
            "        gene_count + not_gene_count AS total_count\n",
            "    \n",
            "      FROM \n",
            "        (SELECT COUNT(*) AS gene_count FROM gene_list_query),\n",
            "        (SELECT COUNT(*) AS not_gene_count FROM not_gene_list_query)\n",
            "    ),\n",
            "\n",
            "    observed_query AS (\n",
            "      -- Table with observed values per pathway in the contingency matrix\n",
            "      SELECT\n",
            "        gene_pp_query.num_genes AS in_gene_in_pathway,\n",
            "        not_gene_pp_query.num_not_genes AS not_gene_in_pathway,\n",
            "        gene_count_query.gene_count - gene_pp_query.num_genes AS in_gene_not_pathway,\n",
            "        gene_count_query.not_gene_count - not_gene_pp_query.num_not_genes AS not_gene_not_pathway,\n",
            "        gene_pp_query.stable_id,\n",
            "        gene_pp_query.name\n",
            "    \n",
            "      FROM \n",
            "        gene_pp_query,\n",
            "        gene_count_query\n",
            "    \n",
            "      INNER JOIN not_gene_pp_query\n",
            "        ON gene_pp_query.stable_id = not_gene_pp_query.stable_id\n",
            "    ),\n",
            "\n",
            "    sum_query AS (\n",
            "      -- Table with summed observed values per pathway in the contingency matrix\n",
            "      SELECT\n",
            "        observed_query.in_gene_in_pathway + observed_query.not_gene_in_pathway AS pathway_total,\n",
            "        observed_query.in_gene_not_pathway + observed_query.not_gene_not_pathway AS not_pathway_total,\n",
            "        observed_query.in_gene_in_pathway + observed_query.in_gene_not_pathway AS gene_total,\n",
            "        observed_query.not_gene_in_pathway + observed_query.not_gene_not_pathway AS not_gene_total,\n",
            "        observed_query.stable_id,\n",
            "        observed_query.name\n",
            "    \n",
            "      FROM\n",
            "        observed_query\n",
            "    ),\n",
            "\n",
            "    expected_query AS (\n",
            "      -- Table with the expected values per pathway in the contingency matrix\n",
            "      SELECT \n",
            "        sum_query.gene_total * sum_query.pathway_total / gene_count_query.total_count AS exp_in_gene_in_pathway,\n",
            "        sum_query.not_gene_total * sum_query.pathway_total / gene_count_query.total_count AS exp_not_gene_in_pathway,\n",
            "        sum_query.gene_total * sum_query.not_pathway_total / gene_count_query.total_count AS exp_in_gene_not_pathway,\n",
            "        sum_query.not_gene_total * sum_query.not_pathway_total / gene_count_query.total_count AS exp_not_gene_not_pathway,\n",
            "        sum_query.stable_id,\n",
            "        sum_query.name\n",
            "    \n",
            "      FROM \n",
            "        sum_query, gene_count_query\n",
            "    ),\n",
            "    \n",
            "    chi_squared_query AS (\n",
            "      -- Table with the chi-squared statistic for each pathway\n",
            "      SELECT\n",
            "        -- Chi squared statistic with Yates' correction\n",
            "        POW(ABS(observed_query.in_gene_in_pathway - expected_query.exp_in_gene_in_pathway) - 0.5, 2) / expected_query.exp_in_gene_in_pathway \n",
            "        + POW(ABS(observed_query.not_gene_in_pathway - expected_query.exp_not_gene_in_pathway) - 0.5, 2) / expected_query.exp_not_gene_in_pathway\n",
            "        + POW(ABS(observed_query.in_gene_not_pathway - expected_query.exp_in_gene_not_pathway) - 0.5, 2) / expected_query.exp_in_gene_not_pathway\n",
            "        + POW(ABS(observed_query.not_gene_not_pathway - expected_query.exp_not_gene_not_pathway) - 0.5, 2) / expected_query.exp_not_gene_not_pathway\n",
            "        AS chi_squared_stat,\n",
            "        observed_query.stable_id,\n",
            "        observed_query.name\n",
            "    \n",
            "      FROM observed_query\n",
            "    \n",
            "      INNER JOIN expected_query\n",
            "        ON observed_query.stable_id = expected_query.stable_id\n",
            "    )\n",
            "\n",
            "  SELECT\n",
            "    observed_query.in_gene_in_pathway,\n",
            "    observed_query.in_gene_not_pathway,\n",
            "    observed_query.not_gene_in_pathway,\n",
            "    observed_query.not_gene_not_pathway,\n",
            "    chi_squared_query.chi_squared_stat,\n",
            "    chi_squared_query.stable_id,\n",
            "    chi_squared_query.name\n",
            "\n",
            "  FROM chi_squared_query\n",
            "\n",
            "  INNER JOIN observed_query\n",
            "    ON chi_squared_query.stable_id = observed_query.stable_id\n",
            "  \n",
            "  INNER JOIN `isb-cgc-bq.reactome_versioned.pathway_v77` AS pathway\n",
            "    ON chi_squared_query.stable_id = pathway.stable_id\n",
            "\n",
            "  WHERE pathway.lowest_level = TRUE\n",
            "\n",
            "  ORDER BY chi_squared_stat DESC\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Fjkbq4FR19in"
      },
      "source": [
        "## Execute the Query\n",
        "\n",
        "Now execute the query to calculate a chi-squared statistic for each pathway:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ua5oYRFY1aVu"
      },
      "source": [
        "# run query and put results in data frame\n",
        "chi_squared_pathways = client.query(final_query).result().to_dataframe()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9Yt-GatY2awc",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "3c62ac76-7a84-4ae3-89a2-9ea7e35f2737"
      },
      "source": [
        "# display the data frame\n",
        "chi_squared_pathways"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>in_gene_in_pathway</th>\n",
              "      <th>in_gene_not_pathway</th>\n",
              "      <th>not_gene_in_pathway</th>\n",
              "      <th>not_gene_not_pathway</th>\n",
              "      <th>chi_squared_stat</th>\n",
              "      <th>stable_id</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>31</td>\n",
              "      <td>11114</td>\n",
              "      <td>33.477023</td>\n",
              "      <td>R-HSA-68962</td>\n",
              "      <td>Activation of the pre-replicative complex</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>3</td>\n",
              "      <td>11142</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>R-HSA-1237112</td>\n",
              "      <td>Methionine salvage pathway</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>3</td>\n",
              "      <td>11142</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>R-HSA-9013425</td>\n",
              "      <td>RHOT1 GTPase cycle</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>50</td>\n",
              "      <td>11095</td>\n",
              "      <td>20.236790</td>\n",
              "      <td>R-HSA-72695</td>\n",
              "      <td>Formation of the ternary complex, and subseque...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>6</td>\n",
              "      <td>11139</td>\n",
              "      <td>18.048197</td>\n",
              "      <td>R-HSA-163765</td>\n",
              "      <td>ChREBP activates metabolic gene expression</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>57</td>\n",
              "      <td>11088</td>\n",
              "      <td>17.513505</td>\n",
              "      <td>R-HSA-72649</td>\n",
              "      <td>Translation initiation complex formation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>57</td>\n",
              "      <td>11088</td>\n",
              "      <td>17.513505</td>\n",
              "      <td>R-HSA-72702</td>\n",
              "      <td>Ribosomal scanning and start codon recognition</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>7</td>\n",
              "      <td>11138</td>\n",
              "      <td>15.671798</td>\n",
              "      <td>R-HSA-8850843</td>\n",
              "      <td>Phosphate bond hydrolysis by NTPDase proteins</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>7</td>\n",
              "      <td>11138</td>\n",
              "      <td>15.671798</td>\n",
              "      <td>R-HSA-68952</td>\n",
              "      <td>DNA replication initiation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>10</td>\n",
              "      <td>11135</td>\n",
              "      <td>11.137000</td>\n",
              "      <td>R-HSA-418885</td>\n",
              "      <td>DCC mediated attractive signaling</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>88</td>\n",
              "      <td>11057</td>\n",
              "      <td>10.567006</td>\n",
              "      <td>R-HSA-192823</td>\n",
              "      <td>Viral mRNA Translation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>92</td>\n",
              "      <td>11053</td>\n",
              "      <td>10.006894</td>\n",
              "      <td>R-HSA-1799339</td>\n",
              "      <td>SRP-dependent cotranslational protein targetin...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>94</td>\n",
              "      <td>11051</td>\n",
              "      <td>9.744550</td>\n",
              "      <td>R-HSA-975956</td>\n",
              "      <td>Nonsense Mediated Decay (NMD) independent of t...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>100</td>\n",
              "      <td>11045</td>\n",
              "      <td>9.020030</td>\n",
              "      <td>R-HSA-72689</td>\n",
              "      <td>Formation of a pool of free 40S subunits</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>110</td>\n",
              "      <td>11035</td>\n",
              "      <td>7.987388</td>\n",
              "      <td>R-HSA-156827</td>\n",
              "      <td>L13a-mediated translational silencing of Cerul...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>111</td>\n",
              "      <td>11034</td>\n",
              "      <td>7.894339</td>\n",
              "      <td>R-HSA-72706</td>\n",
              "      <td>GTP hydrolysis and joining of the 60S ribosoma...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>115</td>\n",
              "      <td>11030</td>\n",
              "      <td>7.538335</td>\n",
              "      <td>R-HSA-975957</td>\n",
              "      <td>Nonsense Mediated Decay (NMD) enhanced by the ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>15</td>\n",
              "      <td>11130</td>\n",
              "      <td>7.362504</td>\n",
              "      <td>R-HSA-204174</td>\n",
              "      <td>Regulation of pyruvate dehydrogenase (PDH) com...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>119</td>\n",
              "      <td>11026</td>\n",
              "      <td>7.206312</td>\n",
              "      <td>R-HSA-114608</td>\n",
              "      <td>Platelet degranulation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>20</td>\n",
              "      <td>11125</td>\n",
              "      <td>5.389681</td>\n",
              "      <td>R-HSA-5651801</td>\n",
              "      <td>PCNA-Dependent Long Patch Base Excision Repair</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>23</td>\n",
              "      <td>11122</td>\n",
              "      <td>4.602357</td>\n",
              "      <td>R-HSA-203927</td>\n",
              "      <td>MicroRNA (miRNA) biogenesis</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-9660826</td>\n",
              "      <td>Purinergic signaling in leishmaniasis infection</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-210991</td>\n",
              "      <td>Basigin interactions</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-5223345</td>\n",
              "      <td>Miscellaneous transport and binding events</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-5696397</td>\n",
              "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>181</td>\n",
              "      <td>10964</td>\n",
              "      <td>3.953753</td>\n",
              "      <td>R-HSA-6791226</td>\n",
              "      <td>Major pathway of rRNA processing in the nucleo...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>29</td>\n",
              "      <td>11116</td>\n",
              "      <td>3.503357</td>\n",
              "      <td>R-HSA-110314</td>\n",
              "      <td>Recognition of DNA damage by PCNA-containing r...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>31</td>\n",
              "      <td>11114</td>\n",
              "      <td>3.229512</td>\n",
              "      <td>R-HSA-5656169</td>\n",
              "      <td>Termination of translesion DNA synthesis</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>33</td>\n",
              "      <td>11112</td>\n",
              "      <td>2.988310</td>\n",
              "      <td>R-HSA-352230</td>\n",
              "      <td>Amino acid transport across the plasma membrane</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>36</td>\n",
              "      <td>11109</td>\n",
              "      <td>2.676134</td>\n",
              "      <td>R-HSA-176187</td>\n",
              "      <td>Activation of ATR in response to replication s...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>36</td>\n",
              "      <td>11109</td>\n",
              "      <td>2.676134</td>\n",
              "      <td>R-HSA-8950505</td>\n",
              "      <td>Gene and protein expression by JAK-STAT signal...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>37</td>\n",
              "      <td>11108</td>\n",
              "      <td>2.583220</td>\n",
              "      <td>R-HSA-5083635</td>\n",
              "      <td>Defective B3GALTL causes Peters-plus syndrome ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>38</td>\n",
              "      <td>11107</td>\n",
              "      <td>2.495164</td>\n",
              "      <td>R-HSA-5173214</td>\n",
              "      <td>O-glycosylation of TSR domain-containing proteins</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>40</td>\n",
              "      <td>11105</td>\n",
              "      <td>2.332203</td>\n",
              "      <td>R-HSA-5696400</td>\n",
              "      <td>Dual Incision in GG-NER</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>54</td>\n",
              "      <td>11091</td>\n",
              "      <td>1.530727</td>\n",
              "      <td>R-HSA-9013409</td>\n",
              "      <td>RHOJ GTPase cycle</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>54</td>\n",
              "      <td>11091</td>\n",
              "      <td>1.530727</td>\n",
              "      <td>R-HSA-193648</td>\n",
              "      <td>NRAGE signals death through JNK</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>63</td>\n",
              "      <td>11082</td>\n",
              "      <td>1.206546</td>\n",
              "      <td>R-HSA-6782210</td>\n",
              "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>64</td>\n",
              "      <td>11081</td>\n",
              "      <td>1.176348</td>\n",
              "      <td>R-HSA-6782135</td>\n",
              "      <td>Dual incision in TC-NER</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>65</td>\n",
              "      <td>11080</td>\n",
              "      <td>1.147121</td>\n",
              "      <td>R-HSA-8936459</td>\n",
              "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>70</td>\n",
              "      <td>11075</td>\n",
              "      <td>1.014107</td>\n",
              "      <td>R-HSA-68949</td>\n",
              "      <td>Orc1 removal from chromatin</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>40</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>73</td>\n",
              "      <td>11072</td>\n",
              "      <td>0.943520</td>\n",
              "      <td>R-HSA-9013408</td>\n",
              "      <td>RHOG GTPase cycle</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>79</td>\n",
              "      <td>11066</td>\n",
              "      <td>0.819465</td>\n",
              "      <td>R-HSA-416482</td>\n",
              "      <td>G alpha (12/13) signalling events</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>80</td>\n",
              "      <td>11065</td>\n",
              "      <td>0.800735</td>\n",
              "      <td>R-HSA-216083</td>\n",
              "      <td>Integrin cell surface interactions</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>87</td>\n",
              "      <td>11058</td>\n",
              "      <td>0.682709</td>\n",
              "      <td>R-HSA-9013404</td>\n",
              "      <td>RAC2 GTPase cycle</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>44</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>91</td>\n",
              "      <td>11054</td>\n",
              "      <td>0.624208</td>\n",
              "      <td>R-HSA-6804756</td>\n",
              "      <td>Regulation of TP53 Activity through Phosphoryl...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>93</td>\n",
              "      <td>11052</td>\n",
              "      <td>0.597060</td>\n",
              "      <td>R-HSA-9013423</td>\n",
              "      <td>RAC3 GTPase cycle</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>99</td>\n",
              "      <td>11046</td>\n",
              "      <td>0.523015</td>\n",
              "      <td>R-HSA-6811434</td>\n",
              "      <td>COPI-dependent Golgi-to-ER retrograde traffic</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>101</td>\n",
              "      <td>11044</td>\n",
              "      <td>0.500559</td>\n",
              "      <td>R-HSA-6807878</td>\n",
              "      <td>COPI-mediated anterograde transport</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>149</td>\n",
              "      <td>10996</td>\n",
              "      <td>0.170910</td>\n",
              "      <td>R-HSA-8980692</td>\n",
              "      <td>RHOA GTPase cycle</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>49</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>158</td>\n",
              "      <td>10987</td>\n",
              "      <td>0.137271</td>\n",
              "      <td>R-HSA-9013148</td>\n",
              "      <td>CDC42 GTPase cycle</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>184</td>\n",
              "      <td>10961</td>\n",
              "      <td>0.067711</td>\n",
              "      <td>R-HSA-9013149</td>\n",
              "      <td>RAC1 GTPase cycle</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>51</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>204</td>\n",
              "      <td>10941</td>\n",
              "      <td>0.034678</td>\n",
              "      <td>R-HSA-5689880</td>\n",
              "      <td>Ub-specific processing proteases</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    in_gene_in_pathway  ...                                               name\n",
              "0                    2  ...          Activation of the pre-replicative complex\n",
              "1                    1  ...                         Methionine salvage pathway\n",
              "2                    1  ...                                 RHOT1 GTPase cycle\n",
              "3                    2  ...  Formation of the ternary complex, and subseque...\n",
              "4                    1  ...         ChREBP activates metabolic gene expression\n",
              "5                    2  ...           Translation initiation complex formation\n",
              "6                    2  ...     Ribosomal scanning and start codon recognition\n",
              "7                    1  ...      Phosphate bond hydrolysis by NTPDase proteins\n",
              "8                    1  ...                         DNA replication initiation\n",
              "9                    1  ...                  DCC mediated attractive signaling\n",
              "10                   2  ...                             Viral mRNA Translation\n",
              "11                   2  ...  SRP-dependent cotranslational protein targetin...\n",
              "12                   2  ...  Nonsense Mediated Decay (NMD) independent of t...\n",
              "13                   2  ...           Formation of a pool of free 40S subunits\n",
              "14                   2  ...  L13a-mediated translational silencing of Cerul...\n",
              "15                   2  ...  GTP hydrolysis and joining of the 60S ribosoma...\n",
              "16                   2  ...  Nonsense Mediated Decay (NMD) enhanced by the ...\n",
              "17                   1  ...  Regulation of pyruvate dehydrogenase (PDH) com...\n",
              "18                   2  ...                            Platelet degranulation \n",
              "19                   1  ...     PCNA-Dependent Long Patch Base Excision Repair\n",
              "20                   1  ...                        MicroRNA (miRNA) biogenesis\n",
              "21                   1  ...    Purinergic signaling in leishmaniasis infection\n",
              "22                   1  ...                               Basigin interactions\n",
              "23                   1  ...         Miscellaneous transport and binding events\n",
              "24                   1  ...  Gap-filling DNA repair synthesis and ligation ...\n",
              "25                   2  ...  Major pathway of rRNA processing in the nucleo...\n",
              "26                   1  ...  Recognition of DNA damage by PCNA-containing r...\n",
              "27                   1  ...           Termination of translesion DNA synthesis\n",
              "28                   1  ...    Amino acid transport across the plasma membrane\n",
              "29                   1  ...  Activation of ATR in response to replication s...\n",
              "30                   1  ...  Gene and protein expression by JAK-STAT signal...\n",
              "31                   1  ...  Defective B3GALTL causes Peters-plus syndrome ...\n",
              "32                   1  ...  O-glycosylation of TSR domain-containing proteins\n",
              "33                   1  ...                            Dual Incision in GG-NER\n",
              "34                   1  ...                                  RHOJ GTPase cycle\n",
              "35                   1  ...                    NRAGE signals death through JNK\n",
              "36                   1  ...  Gap-filling DNA repair synthesis and ligation ...\n",
              "37                   1  ...                            Dual incision in TC-NER\n",
              "38                   1  ...  RUNX1 regulates genes involved in megakaryocyt...\n",
              "39                   1  ...                        Orc1 removal from chromatin\n",
              "40                   1  ...                                  RHOG GTPase cycle\n",
              "41                   1  ...                  G alpha (12/13) signalling events\n",
              "42                   1  ...                 Integrin cell surface interactions\n",
              "43                   1  ...                                  RAC2 GTPase cycle\n",
              "44                   1  ...  Regulation of TP53 Activity through Phosphoryl...\n",
              "45                   1  ...                                  RAC3 GTPase cycle\n",
              "46                   1  ...      COPI-dependent Golgi-to-ER retrograde traffic\n",
              "47                   1  ...                COPI-mediated anterograde transport\n",
              "48                   1  ...                                  RHOA GTPase cycle\n",
              "49                   1  ...                                 CDC42 GTPase cycle\n",
              "50                   1  ...                                  RAC1 GTPase cycle\n",
              "51                   1  ...                   Ub-specific processing proteases\n",
              "\n",
              "[52 rows x 7 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 29
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qM42ztRJDgbM"
      },
      "source": [
        "## Calculate P-Values\n",
        "\n",
        "BigQuery does not have statistical functions to calculate p-values, so we use the SciPy stats library and update the data frame with a new p-value column:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Beim_ePN2cs7",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "6f95818a-7c14-451a-8549-eed4e9598983"
      },
      "source": [
        "chi_squared_pathways['p_value'] = 1-stats.chi2.cdf(chi_squared_pathways['chi_squared_stat'], 1)\n",
        "chi_squared_pathways"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>in_gene_in_pathway</th>\n",
              "      <th>in_gene_not_pathway</th>\n",
              "      <th>not_gene_in_pathway</th>\n",
              "      <th>not_gene_not_pathway</th>\n",
              "      <th>chi_squared_stat</th>\n",
              "      <th>stable_id</th>\n",
              "      <th>name</th>\n",
              "      <th>p_value</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>31</td>\n",
              "      <td>11114</td>\n",
              "      <td>33.477023</td>\n",
              "      <td>R-HSA-68962</td>\n",
              "      <td>Activation of the pre-replicative complex</td>\n",
              "      <td>7.211088e-09</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>3</td>\n",
              "      <td>11142</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>R-HSA-1237112</td>\n",
              "      <td>Methionine salvage pathway</td>\n",
              "      <td>1.313007e-08</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>3</td>\n",
              "      <td>11142</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>R-HSA-9013425</td>\n",
              "      <td>RHOT1 GTPase cycle</td>\n",
              "      <td>1.313007e-08</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>50</td>\n",
              "      <td>11095</td>\n",
              "      <td>20.236790</td>\n",
              "      <td>R-HSA-72695</td>\n",
              "      <td>Formation of the ternary complex, and subseque...</td>\n",
              "      <td>6.842431e-06</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>6</td>\n",
              "      <td>11139</td>\n",
              "      <td>18.048197</td>\n",
              "      <td>R-HSA-163765</td>\n",
              "      <td>ChREBP activates metabolic gene expression</td>\n",
              "      <td>2.153825e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>57</td>\n",
              "      <td>11088</td>\n",
              "      <td>17.513505</td>\n",
              "      <td>R-HSA-72649</td>\n",
              "      <td>Translation initiation complex formation</td>\n",
              "      <td>2.852741e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>57</td>\n",
              "      <td>11088</td>\n",
              "      <td>17.513505</td>\n",
              "      <td>R-HSA-72702</td>\n",
              "      <td>Ribosomal scanning and start codon recognition</td>\n",
              "      <td>2.852741e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>7</td>\n",
              "      <td>11138</td>\n",
              "      <td>15.671798</td>\n",
              "      <td>R-HSA-8850843</td>\n",
              "      <td>Phosphate bond hydrolysis by NTPDase proteins</td>\n",
              "      <td>7.533920e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>7</td>\n",
              "      <td>11138</td>\n",
              "      <td>15.671798</td>\n",
              "      <td>R-HSA-68952</td>\n",
              "      <td>DNA replication initiation</td>\n",
              "      <td>7.533920e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>10</td>\n",
              "      <td>11135</td>\n",
              "      <td>11.137000</td>\n",
              "      <td>R-HSA-418885</td>\n",
              "      <td>DCC mediated attractive signaling</td>\n",
              "      <td>8.462266e-04</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>88</td>\n",
              "      <td>11057</td>\n",
              "      <td>10.567006</td>\n",
              "      <td>R-HSA-192823</td>\n",
              "      <td>Viral mRNA Translation</td>\n",
              "      <td>1.151240e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>92</td>\n",
              "      <td>11053</td>\n",
              "      <td>10.006894</td>\n",
              "      <td>R-HSA-1799339</td>\n",
              "      <td>SRP-dependent cotranslational protein targetin...</td>\n",
              "      <td>1.559553e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>94</td>\n",
              "      <td>11051</td>\n",
              "      <td>9.744550</td>\n",
              "      <td>R-HSA-975956</td>\n",
              "      <td>Nonsense Mediated Decay (NMD) independent of t...</td>\n",
              "      <td>1.798552e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>100</td>\n",
              "      <td>11045</td>\n",
              "      <td>9.020030</td>\n",
              "      <td>R-HSA-72689</td>\n",
              "      <td>Formation of a pool of free 40S subunits</td>\n",
              "      <td>2.670370e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>110</td>\n",
              "      <td>11035</td>\n",
              "      <td>7.987388</td>\n",
              "      <td>R-HSA-156827</td>\n",
              "      <td>L13a-mediated translational silencing of Cerul...</td>\n",
              "      <td>4.710432e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>111</td>\n",
              "      <td>11034</td>\n",
              "      <td>7.894339</td>\n",
              "      <td>R-HSA-72706</td>\n",
              "      <td>GTP hydrolysis and joining of the 60S ribosoma...</td>\n",
              "      <td>4.958976e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>115</td>\n",
              "      <td>11030</td>\n",
              "      <td>7.538335</td>\n",
              "      <td>R-HSA-975957</td>\n",
              "      <td>Nonsense Mediated Decay (NMD) enhanced by the ...</td>\n",
              "      <td>6.039983e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>15</td>\n",
              "      <td>11130</td>\n",
              "      <td>7.362504</td>\n",
              "      <td>R-HSA-204174</td>\n",
              "      <td>Regulation of pyruvate dehydrogenase (PDH) com...</td>\n",
              "      <td>6.659799e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>119</td>\n",
              "      <td>11026</td>\n",
              "      <td>7.206312</td>\n",
              "      <td>R-HSA-114608</td>\n",
              "      <td>Platelet degranulation</td>\n",
              "      <td>7.264762e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>20</td>\n",
              "      <td>11125</td>\n",
              "      <td>5.389681</td>\n",
              "      <td>R-HSA-5651801</td>\n",
              "      <td>PCNA-Dependent Long Patch Base Excision Repair</td>\n",
              "      <td>2.025618e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>23</td>\n",
              "      <td>11122</td>\n",
              "      <td>4.602357</td>\n",
              "      <td>R-HSA-203927</td>\n",
              "      <td>MicroRNA (miRNA) biogenesis</td>\n",
              "      <td>3.192804e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-9660826</td>\n",
              "      <td>Purinergic signaling in leishmaniasis infection</td>\n",
              "      <td>3.631620e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-210991</td>\n",
              "      <td>Basigin interactions</td>\n",
              "      <td>3.631620e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-5223345</td>\n",
              "      <td>Miscellaneous transport and binding events</td>\n",
              "      <td>3.631620e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-5696397</td>\n",
              "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
              "      <td>3.631620e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>181</td>\n",
              "      <td>10964</td>\n",
              "      <td>3.953753</td>\n",
              "      <td>R-HSA-6791226</td>\n",
              "      <td>Major pathway of rRNA processing in the nucleo...</td>\n",
              "      <td>4.676696e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>29</td>\n",
              "      <td>11116</td>\n",
              "      <td>3.503357</td>\n",
              "      <td>R-HSA-110314</td>\n",
              "      <td>Recognition of DNA damage by PCNA-containing r...</td>\n",
              "      <td>6.124455e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>31</td>\n",
              "      <td>11114</td>\n",
              "      <td>3.229512</td>\n",
              "      <td>R-HSA-5656169</td>\n",
              "      <td>Termination of translesion DNA synthesis</td>\n",
              "      <td>7.232223e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>33</td>\n",
              "      <td>11112</td>\n",
              "      <td>2.988310</td>\n",
              "      <td>R-HSA-352230</td>\n",
              "      <td>Amino acid transport across the plasma membrane</td>\n",
              "      <td>8.386764e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>36</td>\n",
              "      <td>11109</td>\n",
              "      <td>2.676134</td>\n",
              "      <td>R-HSA-176187</td>\n",
              "      <td>Activation of ATR in response to replication s...</td>\n",
              "      <td>1.018627e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>36</td>\n",
              "      <td>11109</td>\n",
              "      <td>2.676134</td>\n",
              "      <td>R-HSA-8950505</td>\n",
              "      <td>Gene and protein expression by JAK-STAT signal...</td>\n",
              "      <td>1.018627e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>37</td>\n",
              "      <td>11108</td>\n",
              "      <td>2.583220</td>\n",
              "      <td>R-HSA-5083635</td>\n",
              "      <td>Defective B3GALTL causes Peters-plus syndrome ...</td>\n",
              "      <td>1.080017e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>38</td>\n",
              "      <td>11107</td>\n",
              "      <td>2.495164</td>\n",
              "      <td>R-HSA-5173214</td>\n",
              "      <td>O-glycosylation of TSR domain-containing proteins</td>\n",
              "      <td>1.141965e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>40</td>\n",
              "      <td>11105</td>\n",
              "      <td>2.332203</td>\n",
              "      <td>R-HSA-5696400</td>\n",
              "      <td>Dual Incision in GG-NER</td>\n",
              "      <td>1.267224e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>54</td>\n",
              "      <td>11091</td>\n",
              "      <td>1.530727</td>\n",
              "      <td>R-HSA-9013409</td>\n",
              "      <td>RHOJ GTPase cycle</td>\n",
              "      <td>2.160034e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>54</td>\n",
              "      <td>11091</td>\n",
              "      <td>1.530727</td>\n",
              "      <td>R-HSA-193648</td>\n",
              "      <td>NRAGE signals death through JNK</td>\n",
              "      <td>2.160034e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>63</td>\n",
              "      <td>11082</td>\n",
              "      <td>1.206546</td>\n",
              "      <td>R-HSA-6782210</td>\n",
              "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
              "      <td>2.720173e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>64</td>\n",
              "      <td>11081</td>\n",
              "      <td>1.176348</td>\n",
              "      <td>R-HSA-6782135</td>\n",
              "      <td>Dual incision in TC-NER</td>\n",
              "      <td>2.781007e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>65</td>\n",
              "      <td>11080</td>\n",
              "      <td>1.147121</td>\n",
              "      <td>R-HSA-8936459</td>\n",
              "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
              "      <td>2.841526e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>70</td>\n",
              "      <td>11075</td>\n",
              "      <td>1.014107</td>\n",
              "      <td>R-HSA-68949</td>\n",
              "      <td>Orc1 removal from chromatin</td>\n",
              "      <td>3.139209e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>40</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>73</td>\n",
              "      <td>11072</td>\n",
              "      <td>0.943520</td>\n",
              "      <td>R-HSA-9013408</td>\n",
              "      <td>RHOG GTPase cycle</td>\n",
              "      <td>3.313742e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>79</td>\n",
              "      <td>11066</td>\n",
              "      <td>0.819465</td>\n",
              "      <td>R-HSA-416482</td>\n",
              "      <td>G alpha (12/13) signalling events</td>\n",
              "      <td>3.653366e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>80</td>\n",
              "      <td>11065</td>\n",
              "      <td>0.800735</td>\n",
              "      <td>R-HSA-216083</td>\n",
              "      <td>Integrin cell surface interactions</td>\n",
              "      <td>3.708738e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>87</td>\n",
              "      <td>11058</td>\n",
              "      <td>0.682709</td>\n",
              "      <td>R-HSA-9013404</td>\n",
              "      <td>RAC2 GTPase cycle</td>\n",
              "      <td>4.086556e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>44</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>91</td>\n",
              "      <td>11054</td>\n",
              "      <td>0.624208</td>\n",
              "      <td>R-HSA-6804756</td>\n",
              "      <td>Regulation of TP53 Activity through Phosphoryl...</td>\n",
              "      <td>4.294877e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>93</td>\n",
              "      <td>11052</td>\n",
              "      <td>0.597060</td>\n",
              "      <td>R-HSA-9013423</td>\n",
              "      <td>RAC3 GTPase cycle</td>\n",
              "      <td>4.397019e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>99</td>\n",
              "      <td>11046</td>\n",
              "      <td>0.523015</td>\n",
              "      <td>R-HSA-6811434</td>\n",
              "      <td>COPI-dependent Golgi-to-ER retrograde traffic</td>\n",
              "      <td>4.695582e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>101</td>\n",
              "      <td>11044</td>\n",
              "      <td>0.500559</td>\n",
              "      <td>R-HSA-6807878</td>\n",
              "      <td>COPI-mediated anterograde transport</td>\n",
              "      <td>4.792546e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>149</td>\n",
              "      <td>10996</td>\n",
              "      <td>0.170910</td>\n",
              "      <td>R-HSA-8980692</td>\n",
              "      <td>RHOA GTPase cycle</td>\n",
              "      <td>6.793044e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>49</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>158</td>\n",
              "      <td>10987</td>\n",
              "      <td>0.137271</td>\n",
              "      <td>R-HSA-9013148</td>\n",
              "      <td>CDC42 GTPase cycle</td>\n",
              "      <td>7.110095e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>184</td>\n",
              "      <td>10961</td>\n",
              "      <td>0.067711</td>\n",
              "      <td>R-HSA-9013149</td>\n",
              "      <td>RAC1 GTPase cycle</td>\n",
              "      <td>7.946990e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>51</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>204</td>\n",
              "      <td>10941</td>\n",
              "      <td>0.034678</td>\n",
              "      <td>R-HSA-5689880</td>\n",
              "      <td>Ub-specific processing proteases</td>\n",
              "      <td>8.522717e-01</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    in_gene_in_pathway  ...       p_value\n",
              "0                    2  ...  7.211088e-09\n",
              "1                    1  ...  1.313007e-08\n",
              "2                    1  ...  1.313007e-08\n",
              "3                    2  ...  6.842431e-06\n",
              "4                    1  ...  2.153825e-05\n",
              "5                    2  ...  2.852741e-05\n",
              "6                    2  ...  2.852741e-05\n",
              "7                    1  ...  7.533920e-05\n",
              "8                    1  ...  7.533920e-05\n",
              "9                    1  ...  8.462266e-04\n",
              "10                   2  ...  1.151240e-03\n",
              "11                   2  ...  1.559553e-03\n",
              "12                   2  ...  1.798552e-03\n",
              "13                   2  ...  2.670370e-03\n",
              "14                   2  ...  4.710432e-03\n",
              "15                   2  ...  4.958976e-03\n",
              "16                   2  ...  6.039983e-03\n",
              "17                   1  ...  6.659799e-03\n",
              "18                   2  ...  7.264762e-03\n",
              "19                   1  ...  2.025618e-02\n",
              "20                   1  ...  3.192804e-02\n",
              "21                   1  ...  3.631620e-02\n",
              "22                   1  ...  3.631620e-02\n",
              "23                   1  ...  3.631620e-02\n",
              "24                   1  ...  3.631620e-02\n",
              "25                   2  ...  4.676696e-02\n",
              "26                   1  ...  6.124455e-02\n",
              "27                   1  ...  7.232223e-02\n",
              "28                   1  ...  8.386764e-02\n",
              "29                   1  ...  1.018627e-01\n",
              "30                   1  ...  1.018627e-01\n",
              "31                   1  ...  1.080017e-01\n",
              "32                   1  ...  1.141965e-01\n",
              "33                   1  ...  1.267224e-01\n",
              "34                   1  ...  2.160034e-01\n",
              "35                   1  ...  2.160034e-01\n",
              "36                   1  ...  2.720173e-01\n",
              "37                   1  ...  2.781007e-01\n",
              "38                   1  ...  2.841526e-01\n",
              "39                   1  ...  3.139209e-01\n",
              "40                   1  ...  3.313742e-01\n",
              "41                   1  ...  3.653366e-01\n",
              "42                   1  ...  3.708738e-01\n",
              "43                   1  ...  4.086556e-01\n",
              "44                   1  ...  4.294877e-01\n",
              "45                   1  ...  4.397019e-01\n",
              "46                   1  ...  4.695582e-01\n",
              "47                   1  ...  4.792546e-01\n",
              "48                   1  ...  6.793044e-01\n",
              "49                   1  ...  7.110095e-01\n",
              "50                   1  ...  7.946990e-01\n",
              "51                   1  ...  8.522717e-01\n",
              "\n",
              "[52 rows x 8 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 30
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TqPYE1JZDq5L"
      },
      "source": [
        "## Adjust for Multiple Testing\n",
        "\n",
        "Since we're testing multiple pathways, we need to adjust the p-value threshold for significance accordingly. The number of pathways tested depends on whether or not we're considering all pathways, or just the lowest level pathways. That count can be obtained with the following query."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7KziIxfED8kB"
      },
      "source": [
        "# run query and put results in data frame\n",
        "num_pathways_result = client.query((\"\"\"\n",
        "  SELECT\n",
        "    COUNT (*) AS num_pathways\n",
        "  FROM\n",
        "    `isb-cgc-bq.reactome_versioned.pathway_v77` as pathway\n",
        "  {lowest_level_filter}\n",
        "\"\"\").format(\n",
        "  lowest_level_filter=(\"WHERE lowest_level = TRUE\" if lowest_level else \"\")\n",
        ")).result().to_dataframe()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 80
        },
        "id": "E_7ncwlOEmuG",
        "outputId": "2a300402-e3cc-4cdf-f507-536a4ef38fdd"
      },
      "source": [
        "# display data frame\n",
        "num_pathways_result"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>num_pathways</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1773</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   num_pathways\n",
              "0          1773"
            ]
          },
          "metadata": {},
          "execution_count": 32
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7B9__TlA4M9u",
        "outputId": "0ef921bf-8224-4b14-ce6a-d3418ccaea57"
      },
      "source": [
        "# adjust significance threshold for multiple testing, using a p-value of 0.01\n",
        "num_pathways = num_pathways_result['num_pathways'][0]\n",
        "significance_threshold = 0.01/num_pathways\n",
        "print('Significance Threshold: {}'.format(significance_threshold))\n",
        "\n",
        "# find all pathways that meet the significance criterion after adjusting for\n",
        "# multiple testing\n",
        "significant_pathway_index = chi_squared_pathways['p_value']<significance_threshold\n",
        "\n",
        "# list of significant pathways\n",
        "significant_pathways = chi_squared_pathways[significant_pathway_index]"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Significance Threshold: 5.640157924421884e-06\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6uCQV8mQFlbZ"
      },
      "source": [
        "The final result is a list of all pathways in which the targeted proteins are enriched, or over-represented, at a rate higher than random chance. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 142
        },
        "id": "LfQYh_nNF2UR",
        "outputId": "ed1f0e58-694b-4e21-8ca4-61683ab544e3"
      },
      "source": [
        "# display the final data frame\n",
        "significant_pathways"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>in_gene_in_pathway</th>\n",
              "      <th>in_gene_not_pathway</th>\n",
              "      <th>not_gene_in_pathway</th>\n",
              "      <th>not_gene_not_pathway</th>\n",
              "      <th>chi_squared_stat</th>\n",
              "      <th>stable_id</th>\n",
              "      <th>name</th>\n",
              "      <th>p_value</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>31</td>\n",
              "      <td>11114</td>\n",
              "      <td>33.477023</td>\n",
              "      <td>R-HSA-68962</td>\n",
              "      <td>Activation of the pre-replicative complex</td>\n",
              "      <td>7.211088e-09</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>3</td>\n",
              "      <td>11142</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>R-HSA-1237112</td>\n",
              "      <td>Methionine salvage pathway</td>\n",
              "      <td>1.313007e-08</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>3</td>\n",
              "      <td>11142</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>R-HSA-9013425</td>\n",
              "      <td>RHOT1 GTPase cycle</td>\n",
              "      <td>1.313007e-08</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   in_gene_in_pathway  ...       p_value\n",
              "0                   2  ...  7.211088e-09\n",
              "1                   1  ...  1.313007e-08\n",
              "2                   1  ...  1.313007e-08\n",
              "\n",
              "[3 rows x 8 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 34
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "if9v8H2T12cp"
      },
      "source": [
        "The results of this analysis suggest that at least three pathways may be related to the genes identified.  "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "s31t8Ed-7cfK"
      },
      "source": [
        "## Verify Results by Comparing to SciPy Chi-Squared Function\n",
        "\n",
        "We verify these BigQuery results by calculating the same chi-squared statistic using the SciPy package. Comparing the SciPy p-values and statistics to those of the BigQuery-derived results confirms that they are identical."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-BPVmeMF1CGO"
      },
      "source": [
        "# extract observed values from bigquery result\n",
        "observed = chi_squared_pathways[[\n",
        "  'in_gene_in_pathway',\n",
        "  'in_gene_not_pathway',\n",
        "  'not_gene_in_pathway',\n",
        "  'not_gene_not_pathway'\n",
        "]]\n",
        "\n",
        "# calculate the chi-squared statistic using the scipy stats package  \n",
        "chi2_stat = []\n",
        "chi2_pvalue = []\n",
        "for index, row in observed.iterrows():\n",
        "  stat, pvalue, _, _ = stats.chi2_contingency(\n",
        "    np.reshape(np.matrix(row), (2,2)), correction=True\n",
        "  )\n",
        "  chi2_stat.append(stat)\n",
        "  chi2_pvalue.append(pvalue)\n",
        "\n",
        "# add columns to the original data frame\n",
        "chi_squared_pathways['scipy_stat'] = chi2_stat\n",
        "chi_squared_pathways['scipy_p_value'] = chi2_pvalue"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WWgNFm6w36s9",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "612d8812-294b-43a4-98e8-dad78af21f40"
      },
      "source": [
        "# display the updated data frame\n",
        "chi_squared_pathways"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>in_gene_in_pathway</th>\n",
              "      <th>in_gene_not_pathway</th>\n",
              "      <th>not_gene_in_pathway</th>\n",
              "      <th>not_gene_not_pathway</th>\n",
              "      <th>chi_squared_stat</th>\n",
              "      <th>stable_id</th>\n",
              "      <th>name</th>\n",
              "      <th>p_value</th>\n",
              "      <th>scipy_stat</th>\n",
              "      <th>scipy_p_value</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>31</td>\n",
              "      <td>11114</td>\n",
              "      <td>33.477023</td>\n",
              "      <td>R-HSA-68962</td>\n",
              "      <td>Activation of the pre-replicative complex</td>\n",
              "      <td>7.211088e-09</td>\n",
              "      <td>33.477023</td>\n",
              "      <td>7.211088e-09</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>3</td>\n",
              "      <td>11142</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>R-HSA-1237112</td>\n",
              "      <td>Methionine salvage pathway</td>\n",
              "      <td>1.313007e-08</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>1.313007e-08</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>3</td>\n",
              "      <td>11142</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>R-HSA-9013425</td>\n",
              "      <td>RHOT1 GTPase cycle</td>\n",
              "      <td>1.313007e-08</td>\n",
              "      <td>32.311989</td>\n",
              "      <td>1.313007e-08</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>50</td>\n",
              "      <td>11095</td>\n",
              "      <td>20.236790</td>\n",
              "      <td>R-HSA-72695</td>\n",
              "      <td>Formation of the ternary complex, and subseque...</td>\n",
              "      <td>6.842431e-06</td>\n",
              "      <td>20.236790</td>\n",
              "      <td>6.842431e-06</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>6</td>\n",
              "      <td>11139</td>\n",
              "      <td>18.048197</td>\n",
              "      <td>R-HSA-163765</td>\n",
              "      <td>ChREBP activates metabolic gene expression</td>\n",
              "      <td>2.153825e-05</td>\n",
              "      <td>18.048197</td>\n",
              "      <td>2.153825e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>57</td>\n",
              "      <td>11088</td>\n",
              "      <td>17.513505</td>\n",
              "      <td>R-HSA-72649</td>\n",
              "      <td>Translation initiation complex formation</td>\n",
              "      <td>2.852741e-05</td>\n",
              "      <td>17.513505</td>\n",
              "      <td>2.852741e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>57</td>\n",
              "      <td>11088</td>\n",
              "      <td>17.513505</td>\n",
              "      <td>R-HSA-72702</td>\n",
              "      <td>Ribosomal scanning and start codon recognition</td>\n",
              "      <td>2.852741e-05</td>\n",
              "      <td>17.513505</td>\n",
              "      <td>2.852741e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>7</td>\n",
              "      <td>11138</td>\n",
              "      <td>15.671798</td>\n",
              "      <td>R-HSA-8850843</td>\n",
              "      <td>Phosphate bond hydrolysis by NTPDase proteins</td>\n",
              "      <td>7.533920e-05</td>\n",
              "      <td>15.671798</td>\n",
              "      <td>7.533920e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>7</td>\n",
              "      <td>11138</td>\n",
              "      <td>15.671798</td>\n",
              "      <td>R-HSA-68952</td>\n",
              "      <td>DNA replication initiation</td>\n",
              "      <td>7.533920e-05</td>\n",
              "      <td>15.671798</td>\n",
              "      <td>7.533920e-05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>10</td>\n",
              "      <td>11135</td>\n",
              "      <td>11.137000</td>\n",
              "      <td>R-HSA-418885</td>\n",
              "      <td>DCC mediated attractive signaling</td>\n",
              "      <td>8.462266e-04</td>\n",
              "      <td>11.137000</td>\n",
              "      <td>8.462266e-04</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>88</td>\n",
              "      <td>11057</td>\n",
              "      <td>10.567006</td>\n",
              "      <td>R-HSA-192823</td>\n",
              "      <td>Viral mRNA Translation</td>\n",
              "      <td>1.151240e-03</td>\n",
              "      <td>10.567006</td>\n",
              "      <td>1.151240e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>92</td>\n",
              "      <td>11053</td>\n",
              "      <td>10.006894</td>\n",
              "      <td>R-HSA-1799339</td>\n",
              "      <td>SRP-dependent cotranslational protein targetin...</td>\n",
              "      <td>1.559553e-03</td>\n",
              "      <td>10.006894</td>\n",
              "      <td>1.559553e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>94</td>\n",
              "      <td>11051</td>\n",
              "      <td>9.744550</td>\n",
              "      <td>R-HSA-975956</td>\n",
              "      <td>Nonsense Mediated Decay (NMD) independent of t...</td>\n",
              "      <td>1.798552e-03</td>\n",
              "      <td>9.744550</td>\n",
              "      <td>1.798552e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>100</td>\n",
              "      <td>11045</td>\n",
              "      <td>9.020030</td>\n",
              "      <td>R-HSA-72689</td>\n",
              "      <td>Formation of a pool of free 40S subunits</td>\n",
              "      <td>2.670370e-03</td>\n",
              "      <td>9.020030</td>\n",
              "      <td>2.670370e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>110</td>\n",
              "      <td>11035</td>\n",
              "      <td>7.987388</td>\n",
              "      <td>R-HSA-156827</td>\n",
              "      <td>L13a-mediated translational silencing of Cerul...</td>\n",
              "      <td>4.710432e-03</td>\n",
              "      <td>7.987388</td>\n",
              "      <td>4.710432e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>111</td>\n",
              "      <td>11034</td>\n",
              "      <td>7.894339</td>\n",
              "      <td>R-HSA-72706</td>\n",
              "      <td>GTP hydrolysis and joining of the 60S ribosoma...</td>\n",
              "      <td>4.958976e-03</td>\n",
              "      <td>7.894339</td>\n",
              "      <td>4.958976e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>115</td>\n",
              "      <td>11030</td>\n",
              "      <td>7.538335</td>\n",
              "      <td>R-HSA-975957</td>\n",
              "      <td>Nonsense Mediated Decay (NMD) enhanced by the ...</td>\n",
              "      <td>6.039983e-03</td>\n",
              "      <td>7.538335</td>\n",
              "      <td>6.039983e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>15</td>\n",
              "      <td>11130</td>\n",
              "      <td>7.362504</td>\n",
              "      <td>R-HSA-204174</td>\n",
              "      <td>Regulation of pyruvate dehydrogenase (PDH) com...</td>\n",
              "      <td>6.659799e-03</td>\n",
              "      <td>7.362504</td>\n",
              "      <td>6.659799e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>119</td>\n",
              "      <td>11026</td>\n",
              "      <td>7.206312</td>\n",
              "      <td>R-HSA-114608</td>\n",
              "      <td>Platelet degranulation</td>\n",
              "      <td>7.264762e-03</td>\n",
              "      <td>7.206312</td>\n",
              "      <td>7.264762e-03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>20</td>\n",
              "      <td>11125</td>\n",
              "      <td>5.389681</td>\n",
              "      <td>R-HSA-5651801</td>\n",
              "      <td>PCNA-Dependent Long Patch Base Excision Repair</td>\n",
              "      <td>2.025618e-02</td>\n",
              "      <td>5.389681</td>\n",
              "      <td>2.025618e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>23</td>\n",
              "      <td>11122</td>\n",
              "      <td>4.602357</td>\n",
              "      <td>R-HSA-203927</td>\n",
              "      <td>MicroRNA (miRNA) biogenesis</td>\n",
              "      <td>3.192804e-02</td>\n",
              "      <td>4.602357</td>\n",
              "      <td>3.192804e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-9660826</td>\n",
              "      <td>Purinergic signaling in leishmaniasis infection</td>\n",
              "      <td>3.631620e-02</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>3.631620e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-210991</td>\n",
              "      <td>Basigin interactions</td>\n",
              "      <td>3.631620e-02</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>3.631620e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-5223345</td>\n",
              "      <td>Miscellaneous transport and binding events</td>\n",
              "      <td>3.631620e-02</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>3.631620e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>24</td>\n",
              "      <td>11121</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>R-HSA-5696397</td>\n",
              "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
              "      <td>3.631620e-02</td>\n",
              "      <td>4.382195</td>\n",
              "      <td>3.631620e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>181</td>\n",
              "      <td>10964</td>\n",
              "      <td>3.953753</td>\n",
              "      <td>R-HSA-6791226</td>\n",
              "      <td>Major pathway of rRNA processing in the nucleo...</td>\n",
              "      <td>4.676696e-02</td>\n",
              "      <td>3.953753</td>\n",
              "      <td>4.676696e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>29</td>\n",
              "      <td>11116</td>\n",
              "      <td>3.503357</td>\n",
              "      <td>R-HSA-110314</td>\n",
              "      <td>Recognition of DNA damage by PCNA-containing r...</td>\n",
              "      <td>6.124455e-02</td>\n",
              "      <td>3.503357</td>\n",
              "      <td>6.124455e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>31</td>\n",
              "      <td>11114</td>\n",
              "      <td>3.229512</td>\n",
              "      <td>R-HSA-5656169</td>\n",
              "      <td>Termination of translesion DNA synthesis</td>\n",
              "      <td>7.232223e-02</td>\n",
              "      <td>3.229512</td>\n",
              "      <td>7.232223e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>33</td>\n",
              "      <td>11112</td>\n",
              "      <td>2.988310</td>\n",
              "      <td>R-HSA-352230</td>\n",
              "      <td>Amino acid transport across the plasma membrane</td>\n",
              "      <td>8.386764e-02</td>\n",
              "      <td>2.988310</td>\n",
              "      <td>8.386764e-02</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>36</td>\n",
              "      <td>11109</td>\n",
              "      <td>2.676134</td>\n",
              "      <td>R-HSA-176187</td>\n",
              "      <td>Activation of ATR in response to replication s...</td>\n",
              "      <td>1.018627e-01</td>\n",
              "      <td>2.676134</td>\n",
              "      <td>1.018627e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>36</td>\n",
              "      <td>11109</td>\n",
              "      <td>2.676134</td>\n",
              "      <td>R-HSA-8950505</td>\n",
              "      <td>Gene and protein expression by JAK-STAT signal...</td>\n",
              "      <td>1.018627e-01</td>\n",
              "      <td>2.676134</td>\n",
              "      <td>1.018627e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>37</td>\n",
              "      <td>11108</td>\n",
              "      <td>2.583220</td>\n",
              "      <td>R-HSA-5083635</td>\n",
              "      <td>Defective B3GALTL causes Peters-plus syndrome ...</td>\n",
              "      <td>1.080017e-01</td>\n",
              "      <td>2.583220</td>\n",
              "      <td>1.080017e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>38</td>\n",
              "      <td>11107</td>\n",
              "      <td>2.495164</td>\n",
              "      <td>R-HSA-5173214</td>\n",
              "      <td>O-glycosylation of TSR domain-containing proteins</td>\n",
              "      <td>1.141965e-01</td>\n",
              "      <td>2.495164</td>\n",
              "      <td>1.141965e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>40</td>\n",
              "      <td>11105</td>\n",
              "      <td>2.332203</td>\n",
              "      <td>R-HSA-5696400</td>\n",
              "      <td>Dual Incision in GG-NER</td>\n",
              "      <td>1.267224e-01</td>\n",
              "      <td>2.332203</td>\n",
              "      <td>1.267224e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>54</td>\n",
              "      <td>11091</td>\n",
              "      <td>1.530727</td>\n",
              "      <td>R-HSA-9013409</td>\n",
              "      <td>RHOJ GTPase cycle</td>\n",
              "      <td>2.160034e-01</td>\n",
              "      <td>1.530727</td>\n",
              "      <td>2.160034e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>54</td>\n",
              "      <td>11091</td>\n",
              "      <td>1.530727</td>\n",
              "      <td>R-HSA-193648</td>\n",
              "      <td>NRAGE signals death through JNK</td>\n",
              "      <td>2.160034e-01</td>\n",
              "      <td>1.530727</td>\n",
              "      <td>2.160034e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>63</td>\n",
              "      <td>11082</td>\n",
              "      <td>1.206546</td>\n",
              "      <td>R-HSA-6782210</td>\n",
              "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
              "      <td>2.720173e-01</td>\n",
              "      <td>1.206546</td>\n",
              "      <td>2.720173e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>64</td>\n",
              "      <td>11081</td>\n",
              "      <td>1.176348</td>\n",
              "      <td>R-HSA-6782135</td>\n",
              "      <td>Dual incision in TC-NER</td>\n",
              "      <td>2.781007e-01</td>\n",
              "      <td>1.176348</td>\n",
              "      <td>2.781007e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>65</td>\n",
              "      <td>11080</td>\n",
              "      <td>1.147121</td>\n",
              "      <td>R-HSA-8936459</td>\n",
              "      <td>RUNX1 regulates genes involved in megakaryocyt...</td>\n",
              "      <td>2.841526e-01</td>\n",
              "      <td>1.147121</td>\n",
              "      <td>2.841526e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>70</td>\n",
              "      <td>11075</td>\n",
              "      <td>1.014107</td>\n",
              "      <td>R-HSA-68949</td>\n",
              "      <td>Orc1 removal from chromatin</td>\n",
              "      <td>3.139209e-01</td>\n",
              "      <td>1.014107</td>\n",
              "      <td>3.139209e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>40</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>73</td>\n",
              "      <td>11072</td>\n",
              "      <td>0.943520</td>\n",
              "      <td>R-HSA-9013408</td>\n",
              "      <td>RHOG GTPase cycle</td>\n",
              "      <td>3.313742e-01</td>\n",
              "      <td>0.943520</td>\n",
              "      <td>3.313742e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>79</td>\n",
              "      <td>11066</td>\n",
              "      <td>0.819465</td>\n",
              "      <td>R-HSA-416482</td>\n",
              "      <td>G alpha (12/13) signalling events</td>\n",
              "      <td>3.653366e-01</td>\n",
              "      <td>0.819465</td>\n",
              "      <td>3.653366e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>80</td>\n",
              "      <td>11065</td>\n",
              "      <td>0.800735</td>\n",
              "      <td>R-HSA-216083</td>\n",
              "      <td>Integrin cell surface interactions</td>\n",
              "      <td>3.708738e-01</td>\n",
              "      <td>0.800735</td>\n",
              "      <td>3.708738e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>87</td>\n",
              "      <td>11058</td>\n",
              "      <td>0.682709</td>\n",
              "      <td>R-HSA-9013404</td>\n",
              "      <td>RAC2 GTPase cycle</td>\n",
              "      <td>4.086556e-01</td>\n",
              "      <td>0.682709</td>\n",
              "      <td>4.086556e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>44</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>91</td>\n",
              "      <td>11054</td>\n",
              "      <td>0.624208</td>\n",
              "      <td>R-HSA-6804756</td>\n",
              "      <td>Regulation of TP53 Activity through Phosphoryl...</td>\n",
              "      <td>4.294877e-01</td>\n",
              "      <td>0.624208</td>\n",
              "      <td>4.294877e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>93</td>\n",
              "      <td>11052</td>\n",
              "      <td>0.597060</td>\n",
              "      <td>R-HSA-9013423</td>\n",
              "      <td>RAC3 GTPase cycle</td>\n",
              "      <td>4.397019e-01</td>\n",
              "      <td>0.597060</td>\n",
              "      <td>4.397019e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>99</td>\n",
              "      <td>11046</td>\n",
              "      <td>0.523015</td>\n",
              "      <td>R-HSA-6811434</td>\n",
              "      <td>COPI-dependent Golgi-to-ER retrograde traffic</td>\n",
              "      <td>4.695582e-01</td>\n",
              "      <td>0.523015</td>\n",
              "      <td>4.695582e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>101</td>\n",
              "      <td>11044</td>\n",
              "      <td>0.500559</td>\n",
              "      <td>R-HSA-6807878</td>\n",
              "      <td>COPI-mediated anterograde transport</td>\n",
              "      <td>4.792546e-01</td>\n",
              "      <td>0.500559</td>\n",
              "      <td>4.792546e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>149</td>\n",
              "      <td>10996</td>\n",
              "      <td>0.170910</td>\n",
              "      <td>R-HSA-8980692</td>\n",
              "      <td>RHOA GTPase cycle</td>\n",
              "      <td>6.793044e-01</td>\n",
              "      <td>0.170910</td>\n",
              "      <td>6.793044e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>49</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>158</td>\n",
              "      <td>10987</td>\n",
              "      <td>0.137271</td>\n",
              "      <td>R-HSA-9013148</td>\n",
              "      <td>CDC42 GTPase cycle</td>\n",
              "      <td>7.110095e-01</td>\n",
              "      <td>0.137271</td>\n",
              "      <td>7.110095e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>184</td>\n",
              "      <td>10961</td>\n",
              "      <td>0.067711</td>\n",
              "      <td>R-HSA-9013149</td>\n",
              "      <td>RAC1 GTPase cycle</td>\n",
              "      <td>7.946990e-01</td>\n",
              "      <td>0.067711</td>\n",
              "      <td>7.946990e-01</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>51</th>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "      <td>204</td>\n",
              "      <td>10941</td>\n",
              "      <td>0.034678</td>\n",
              "      <td>R-HSA-5689880</td>\n",
              "      <td>Ub-specific processing proteases</td>\n",
              "      <td>8.522717e-01</td>\n",
              "      <td>0.034678</td>\n",
              "      <td>8.522717e-01</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    in_gene_in_pathway  in_gene_not_pathway  ...  scipy_stat  scipy_p_value\n",
              "0                    2                   19  ...   33.477023   7.211088e-09\n",
              "1                    1                   20  ...   32.311989   1.313007e-08\n",
              "2                    1                   20  ...   32.311989   1.313007e-08\n",
              "3                    2                   19  ...   20.236790   6.842431e-06\n",
              "4                    1                   20  ...   18.048197   2.153825e-05\n",
              "5                    2                   19  ...   17.513505   2.852741e-05\n",
              "6                    2                   19  ...   17.513505   2.852741e-05\n",
              "7                    1                   20  ...   15.671798   7.533920e-05\n",
              "8                    1                   20  ...   15.671798   7.533920e-05\n",
              "9                    1                   20  ...   11.137000   8.462266e-04\n",
              "10                   2                   19  ...   10.567006   1.151240e-03\n",
              "11                   2                   19  ...   10.006894   1.559553e-03\n",
              "12                   2                   19  ...    9.744550   1.798552e-03\n",
              "13                   2                   19  ...    9.020030   2.670370e-03\n",
              "14                   2                   19  ...    7.987388   4.710432e-03\n",
              "15                   2                   19  ...    7.894339   4.958976e-03\n",
              "16                   2                   19  ...    7.538335   6.039983e-03\n",
              "17                   1                   20  ...    7.362504   6.659799e-03\n",
              "18                   2                   19  ...    7.206312   7.264762e-03\n",
              "19                   1                   20  ...    5.389681   2.025618e-02\n",
              "20                   1                   20  ...    4.602357   3.192804e-02\n",
              "21                   1                   20  ...    4.382195   3.631620e-02\n",
              "22                   1                   20  ...    4.382195   3.631620e-02\n",
              "23                   1                   20  ...    4.382195   3.631620e-02\n",
              "24                   1                   20  ...    4.382195   3.631620e-02\n",
              "25                   2                   19  ...    3.953753   4.676696e-02\n",
              "26                   1                   20  ...    3.503357   6.124455e-02\n",
              "27                   1                   20  ...    3.229512   7.232223e-02\n",
              "28                   1                   20  ...    2.988310   8.386764e-02\n",
              "29                   1                   20  ...    2.676134   1.018627e-01\n",
              "30                   1                   20  ...    2.676134   1.018627e-01\n",
              "31                   1                   20  ...    2.583220   1.080017e-01\n",
              "32                   1                   20  ...    2.495164   1.141965e-01\n",
              "33                   1                   20  ...    2.332203   1.267224e-01\n",
              "34                   1                   20  ...    1.530727   2.160034e-01\n",
              "35                   1                   20  ...    1.530727   2.160034e-01\n",
              "36                   1                   20  ...    1.206546   2.720173e-01\n",
              "37                   1                   20  ...    1.176348   2.781007e-01\n",
              "38                   1                   20  ...    1.147121   2.841526e-01\n",
              "39                   1                   20  ...    1.014107   3.139209e-01\n",
              "40                   1                   20  ...    0.943520   3.313742e-01\n",
              "41                   1                   20  ...    0.819465   3.653366e-01\n",
              "42                   1                   20  ...    0.800735   3.708738e-01\n",
              "43                   1                   20  ...    0.682709   4.086556e-01\n",
              "44                   1                   20  ...    0.624208   4.294877e-01\n",
              "45                   1                   20  ...    0.597060   4.397019e-01\n",
              "46                   1                   20  ...    0.523015   4.695582e-01\n",
              "47                   1                   20  ...    0.500559   4.792546e-01\n",
              "48                   1                   20  ...    0.170910   6.793044e-01\n",
              "49                   1                   20  ...    0.137271   7.110095e-01\n",
              "50                   1                   20  ...    0.067711   7.946990e-01\n",
              "51                   1                   20  ...    0.034678   8.522717e-01\n",
              "\n",
              "[52 rows x 10 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 36
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HBGSd1wS5GRK"
      },
      "source": [
        "# Conclusion\n",
        "\n",
        "This notebook demonstrated usage of the Reactome BigQuery dataset for basic cancer pathway identification from a gene set, as well as a more complex pathway enrichment analysis using a chi-squared statistic."
      ]
    }
  ]
}
