{
  "cells": [
    {
      "cell_type": "markdown",
      "id": "08c3e7f4",
      "metadata": {
        "id": "08c3e7f4"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/isb-cgc/Community-Notebooks/blob/jacob-dev/MitelmanDB/Mitelman_DB_Queries.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "8abfde41",
      "metadata": {
        "id": "8abfde41"
      },
      "source": [
        "# Navigating the Mitelman Database in BigQuery Using SQL\n",
        "\n",
        "Check out other notebooks at our [Community Notebooks Repository](https://github.com/isb-cgc/Community-Notebooks)!\n",
        "\n",
        "- Title: Navigating the Mitelman Database in BigQuery Using SQL\n",
        "- Author: Jacob Wilson\n",
        "- Created: 2025-01-01\n",
        "- URL: https://github.com/isb-cgc/Community-Notebooks/blob/jacob-dev/MitelmanDB/Mitelman_DB_Queries.ipynb\n",
        "- Purpose: Introduce the Mitelman Database BigQuery Tables and how to obtain information using basic SQL queries.\n",
        "<br/>\n",
        "\n",
        "In this notebook, we will use SQL queries to obtain various types of data available in the Mitelman Database BigQuery tables."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "UpvufWbZeovQ",
      "metadata": {
        "id": "UpvufWbZeovQ"
      },
      "source": [
        "## BigQuery and SQL Background\n",
        "BigQuery is a data warehouse platform that is available from Google. It provides a powerful way of storing and analyzing tabular data. We use this platform heavily in ISB-CGC to store our data and provide our users with extensive data analysis capabilities.\n",
        "\n",
        "Users interact with data in BigQuery tables using SQL-like queries in the Google Cloud Console. SQL (Structured Query Language) is a common language that is used for many databases. It provides an easy to understand syntax that is used for working with database tables.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Mitelman Database Table Descriptions"
      ],
      "metadata": {
        "id": "TofyX0jfia04"
      },
      "id": "TofyX0jfia04"
    },
    {
      "cell_type": "markdown",
      "source": [
        "|Table              |Data Type         |Table Description    |\n",
        "|:-------------------|:------------------|:---------------------|\n",
        "|AuthorReference    | LITERATURE       | Data was compiled from the original publications used in the Mitelman database. Author names were manually curated for each assigned reference number. The order of the names in the list of authors for the given reference is recorded.                |\n",
        " |CytoBands_hg38     | CYTOGENETIC      | Data was extracted from the UCSC database. Start and end coordinates of each cytoband were obtained for all chromosomes based on genome build hg38.                     |\n",
        "|CytoConverted      | CYTOGENETIC      | Data was processed using karyotype nomenclature from the CytoConverterSample table. Chromosomal imbalances and associated genomic coordinates were generated by the CytoConverter software. More details: see https://github.com/isb-cgc/ISB-CGC-CytoConverter/                |\n",
        "|CytoConvertedLog   | METADATA         | Data was generated by the CytoConverter software. Warnings and errors were returned from the software while processing karyotype nomenclature from the CytoConverterSample table. More details: see https://github.com/isb-cgc/ISB-CGC-CytoConverter/               |\n",
        "|Cytogen            | CLINICAL DATA    | Data was compiled from the original publications used in the Mitelman database. Patient demographics and disease data were manually curated for each assigned reference and case number.               |\n",
        "|CytogenInv         | CYTOGENETIC      | Data was compiled from the original publications used in the Mitelman database. Karyotype nomenclature data were manually curated for each assigned reference, case, and investigation number.                |\n",
        "|CytogenInvValid    | CYTOGENETIC      | Data was generated using the CytogenInv table. Karyotype nomenclature was processed using a syntax checker that validated the string for use with the CytoConverter software.               |\n",
        "|KaryAbnorm         | CYTOGENETIC      | Data was generated using the CytogenInv table. Individual karyotype nomenclature abnormalities are maintained for searching purposes.               |\n",
        "|KaryBit            | CYTOGENETIC      | Data was generated using the CytogenInv table. Includes all elements of karyotype nomenclature when split on commas.               |\n",
        "|KaryBreak          | CYTOGENETIC      | Data was generated using the CytogenInv table. Breakpoints were extracted from the karyotype nomenclature.                |\n",
        "|KaryClone          | CYTOGENETIC      | Data was generated using the CytogenInv table. Contains separated clones from karyotype nomenclature.               |\n",
        "|Koder              | CLINICAL SUPPLEMENT | Data was compiled from the original publications used in the Mitelman database. Various fields of data were assigned a unique code to be used as a reference in other database tables.               |\n",
        "|MolBiolClinAssoc   | CYTOGENETIC      | Data was compiled from the original publications used in the Mitelman database. Gene fusions associated with karyotype nomenclature were manually curated for each reference.               |\n",
        "|MolClinAbnorm      | CYTOGENETIC      | Data was generated using the MolBiolClinAssoc table. Karyotype nomenclature was split into individual abnormalities.               |\n",
        "|MolClinBreak       | CYTOGENETIC      | Data was generated using the MolBiolClinAssoc table. Breakpoints were extracted from the karyotype nomenclature.               |\n",
        "|MolClinGene        | GENE FUSION      | Data was compiled from the original publications used in the Mitelman database. Gene fusions were manually curated for each reference.               |\n",
        "|RecurrentData      | GENE FUSION      | Data was aggregated from the separated chromosomal abnormalities. Frequency counts of each structural abnormality are recorded.                |\n",
        "|RecurrentNumData   | GENE FUSION      | Data was aggregated from the separated chromosomal abnormalities. Frequency counts of each numerical abnormality are recorded.                |\n",
        "|Reference          | LITERATURE       | Data was compiled from the original publications used in the Mitelman database. Information associated with the reviewed publication is recorded and assigned a reference number.               |"
      ],
      "metadata": {
        "id": "GmmL4i49oVaN"
      },
      "id": "GmmL4i49oVaN"
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Potential columns of interest\n"
      ],
      "metadata": {
        "id": "A3pzv_m1PqHe"
      },
      "id": "A3pzv_m1PqHe"
    },
    {
      "cell_type": "markdown",
      "source": [
        "Find cytogenetic abnormalities and gene fusions:\n",
        "\n",
        "| Table              |Column Name         |Description |\n",
        "|:-------------------|:------------------|:--------|\n",
        "|CytoConverted    | Type       |Type of copy number aberration as determined by CytoConverter (gain or loss).|\n",
        "|     | Start      |Start coordinate of the aberration identified by CytoConverter. Based on genome build hg38.|\n",
        "|    | End       |End coordinate of the aberration identified by CytoConverter. Based on genome build hg38.|\n",
        "|     | Chr      |Chromosome number (1- 22, X or Y) in format \"chr##\"|\n",
        "|CytogenInv    | KaryShort       |Short (possibly truncated) karyotype description|\n",
        "|KaryAbnorm     | Abnormality      |Abnormality description, eg 'del(15)(q22q24)', 't(17;17)(p11;q11)', '-22', etc|\n",
        "|KaryBreak     | Breakpoint      |Cytogenetic band location of abnormality, e.g., 1p21, as defined by ISCN.|\n",
        "|KaryClone | ChromoMax |The maximum modal number present in the karyotype nomenclature. |\n",
        "| |ChromoMin |The minimum modal number present in the karyotype nomenclature. |\n",
        "|MolClinGene|Gene |HGNC gene symbol |"
      ],
      "metadata": {
        "id": "xCqT5XE28dkw"
      },
      "id": "xCqT5XE28dkw"
    },
    {
      "cell_type": "markdown",
      "source": [
        "Find information relevant to publications:\n",
        "\n",
        "|Table              |Column Name         |Description |\n",
        "|:-------------------|:------------------|:--------|\n",
        "|AuthorReference |Name |Author name |\n",
        "|Reference |Journal |Journal name abbreviation, eg 'Cancer Genet Cytogenet' |\n",
        "| |Year |Publication year |"
      ],
      "metadata": {
        "id": "vqkdH6x68eKn"
      },
      "id": "vqkdH6x68eKn"
    },
    {
      "cell_type": "markdown",
      "source": [
        "Find information relevant to disease:\n",
        "\n",
        "|Table              |Column Name         |Description |\n",
        "|:-------------------|:------------------|:--------|\n",
        "|Cytogen |Morph |Unique code indicating the morphology type as present in Koder table. |\n",
        "| |Topo |Unique code indicating the topography type as present in Koder table. |\n",
        "|CytogenInv |Tissue |Indicates the type of tissue. \"BM\" means Bone Marrow, \"TB\" means Tumor biopsy, \"LN\" means Lymph node, \"EX\" means Exudate, and \"CSF\" means Cerebrospinal fluid. |"
      ],
      "metadata": {
        "id": "726FjfHq8dYN"
      },
      "id": "726FjfHq8dYN"
    },
    {
      "cell_type": "markdown",
      "source": [
        "Find general information:\n",
        "\n",
        "|Table              |Column Name         |Description |\n",
        "|:-------------------|:------------------|:--------|\n",
        "|Koder |Benamning |The 'long' name corresponding to this code, eg 'Vascular and perivascular tumors (all subtypes)' |\n",
        "| |Kod |Code number which depends on the KodTyp |\n",
        "| |KodType |This field indicates the type of the code -- there are 7 types: 'MORPH' (morphology), 'GEO' (geography), 'TOP' (topography), 'HER' (heredity), 'TISSUE', 'TREAT' (treatment), and 'RACE' |"
      ],
      "metadata": {
        "id": "OMcUOQtF8dHN"
      },
      "id": "OMcUOQtF8dHN"
    },
    {
      "cell_type": "markdown",
      "id": "9e64559c",
      "metadata": {
        "id": "9e64559c"
      },
      "source": [
        "## Initialize Notebook Environment\n",
        "\n",
        "Before beginning, we first need to load dependencies and authenticate to BigQuery."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4a073c7c",
      "metadata": {
        "id": "4a073c7c"
      },
      "source": [
        "## Install Dependencies"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "id": "478a0ce9",
      "metadata": {
        "id": "478a0ce9"
      },
      "outputs": [],
      "source": [
        "# GCP libraries\n",
        "from google.cloud import bigquery\n",
        "from google.colab import auth"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "43816bcc",
      "metadata": {
        "id": "43816bcc"
      },
      "source": [
        "## Authenticate\n",
        "\n",
        "In order to utilize BigQuery, we must obtain authorization to BigQuery and Google Cloud."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "id": "6aa17ca5",
      "metadata": {
        "id": "6aa17ca5"
      },
      "outputs": [],
      "source": [
        "# if you're using Google Colab, authenticate to gcloud with the following\n",
        "auth.authenticate_user()\n",
        "\n",
        "# alternatively, use the gcloud SDK\n",
        "#!gcloud auth application-default login"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6529ac99",
      "metadata": {
        "id": "6529ac99"
      },
      "source": [
        "## Google project ID\n",
        "\n",
        "Set your own Google project ID for use with this notebook."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "id": "cabd29da",
      "metadata": {
        "id": "cabd29da"
      },
      "outputs": [],
      "source": [
        "# set the google project that will be billed for this notebook's computations\n",
        "google_project = 'isb-cgc-notebook-dev'  ## change this"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "9a4fdbc6",
      "metadata": {
        "id": "9a4fdbc6"
      },
      "source": [
        "## BigQuery Client"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "id": "a3235756",
      "metadata": {
        "id": "a3235756"
      },
      "outputs": [],
      "source": [
        "# Initialize a client to access the data within BigQuery\n",
        "if google_project == 'your_project_id':\n",
        "    print('Please update the project ID with your Google Cloud Project')\n",
        "else:\n",
        "    client = bigquery.Client(google_project)\n",
        "\n",
        "# set the Mitelman Database project\n",
        "bq_project = 'mitelman-db'\n",
        "bq_dataset = 'prod'"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "9cdbb0f1",
      "metadata": {
        "id": "9cdbb0f1"
      },
      "source": [
        "# Querying for Information"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Find the column names of the 'Cytogen' table. The 'table_name'\n",
        "# value can be changed to find the column names of other tables.\n",
        "col_names = f'''\n",
        "SELECT column_name\n",
        "FROM `{bq_project}`.{bq_dataset}.INFORMATION_SCHEMA.COLUMNS\n",
        "WHERE table_name = 'Cytogen'\n",
        "ORDER BY column_name\n",
        "'''\n",
        "\n",
        "# Return the distict code types.\n",
        "kod_types = f'''\n",
        "SELECT DISTINCT KodTyp\n",
        "FROM `{bq_project}.{bq_dataset}.Koder`\n",
        "'''\n",
        "\n",
        "# Find the code, full name, and abbreviations for the code type\n",
        "# 'MORPH'.\n",
        "all_morph = f'''\n",
        "SELECT Kod, KodTyp, Benamning AS FullName, Kortnamn AS ShortName\n",
        "FROM `{bq_project}.{bq_dataset}.Koder`\n",
        "WHERE KodTyp = \"MORPH\"\n",
        "ORDER BY Kortnamn\n",
        "'''\n",
        "\n",
        "# Same as above but for the code type 'TOP'\n",
        "all_topo = f'''\n",
        "SELECT Kod, KodTyp, Benamning AS FullName, Kortnamn AS ShortName\n",
        "FROM `{bq_project}.{bq_dataset}.Koder`\n",
        "WHERE KodTyp = \"TOP\"\n",
        "ORDER BY Kortnamn\n",
        "'''"
      ],
      "metadata": {
        "id": "kNsZ3Jt5JX8S"
      },
      "id": "kNsZ3Jt5JX8S",
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "columns = client.query(col_names).result().to_dataframe()\n",
        "codes = client.query(kod_types).result().to_dataframe()\n",
        "morph = client.query(all_morph).result().to_dataframe()\n",
        "topo = client.query(all_topo).result().to_dataframe()"
      ],
      "metadata": {
        "id": "QrFAE094a5td"
      },
      "id": "QrFAE094a5td",
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(columns.head())\n",
        "print(codes.head())\n",
        "print(morph.head())\n",
        "print(topo.head())"
      ],
      "metadata": {
        "id": "WupBeQrlkbcp",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "outputId": "993206df-836a-478f-d83f-cb2c7eae1126"
      },
      "id": "WupBeQrlkbcp",
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  column_name\n",
            "0         Age\n",
            "1      CaseNo\n",
            "2   CaseOrder\n",
            "3     Country\n",
            "4      HerDis\n",
            "  KodTyp\n",
            "0    GEO\n",
            "1    HER\n",
            "2    TOP\n",
            "3   RACE\n",
            "4  MORPH\n",
            "    Kod KodTyp                           FullName ShortName\n",
            "0  8725  MORPH               Aneurysmal bone cyst       ABC\n",
            "1  1115  MORPH          Acute basophilic leukemia       ABL\n",
            "2  1303  MORPH  Atypical chronic myeloid leukemia      ACML\n",
            "3  1117  MORPH        Acute eosinophilic leukemia       AEL\n",
            "4  8506  MORPH   Angiomatoid fibrous histiocytoma       AFH\n",
            "    Kod KodTyp                           FullName      ShortName\n",
            "0  0703    TOP                            Adrenal        Adrenal\n",
            "1  0230    TOP                               Anus           Anus\n",
            "2  1201    TOP                       Blood vessel      Bl vessel\n",
            "3  0305    TOP                            Bladder        Bladder\n",
            "4    09    TOP  Bone and soft tissues (all sites)  Bone and soft\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Using SQL Aggregate Functions\n",
        "\n",
        "Aggregate functions in SQL are useful for generating basic statistics. Commonly used aggregate functions include:\n",
        "- COUNT(): returns the number of rows in a chosen set\n",
        "- MIN(): minimum value in the column\n",
        "- MAX(): maximum value in the column\n",
        "- AVG(): average of values in a column\n",
        "- SUM(): total sum of a column\n",
        "\n",
        "It is often necesary to use the GROUP BY clause with certain aggregate functions to create a set for the function to be applied to."
      ],
      "metadata": {
        "id": "Ls0P5yTmhj2V"
      },
      "id": "Ls0P5yTmhj2V"
    },
    {
      "cell_type": "code",
      "source": [
        "# Count the number of 'M' and 'F' entries in the 'Sex' column\n",
        "count_sex = f'''\n",
        "SELECT\n",
        "SUM(IF(Sex = 'M', 1, 0)) as Male,\n",
        "SUM(IF(Sex = 'F', 1, 0)) As Female\n",
        "FROM `{bq_project}.{bq_dataset}.Cytogen`\n",
        "'''\n",
        "\n",
        "# Select all ages and provide a count for each age\n",
        "age_stats = f'''\n",
        "SELECT Age, COUNT(Age) AS Count\n",
        "FROM `{bq_project}.{bq_dataset}.Cytogen`\n",
        "GROUP BY Age\n",
        "ORDER BY Age\n",
        "'''\n",
        "\n",
        "# Find the minimum, maximum, and average of ages\n",
        "age_range = f'''\n",
        "SELECT MIN(Age) AS min_age, MAX(Age) AS max_age, AVG(Age) AS avg_age\n",
        "FROM `{bq_project}.{bq_dataset}.Cytogen`\n",
        "'''\n",
        "\n",
        "# Find gene fusions and count the number of occurrences\n",
        "query_fusions = f'''\n",
        "SELECT g.Gene, COUNT(g.Gene) AS Count\n",
        "FROM `{bq_project}.{bq_dataset}.MolClinGene` g\n",
        "-- gene name for fusions is double-colon separated gene pair\n",
        "WHERE g.Gene LIKE \"%::%\"\n",
        "GROUP BY g.Gene\n",
        "ORDER BY Count DESC\n",
        "'''"
      ],
      "metadata": {
        "id": "uP4l_C-ThnNb"
      },
      "id": "uP4l_C-ThnNb",
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "counts = client.query(count_sex).result().to_dataframe()\n",
        "ages = client.query(age_stats).result().to_dataframe()\n",
        "age_range = client.query(age_range).result().to_dataframe()\n",
        "fusions = client.query(query_fusions).result().to_dataframe()"
      ],
      "metadata": {
        "id": "JgkCANwya6X6"
      },
      "id": "JgkCANwya6X6",
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print('Number of male and female cases:')\n",
        "print(counts)\n",
        "print('\\nAge distribution:')\n",
        "print(ages)\n",
        "print('\\nAge range:')\n",
        "print(age_range)\n",
        "print('\\nTypes and counts of fusions:')\n",
        "print(fusions)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LuXji4Rgkh_Y",
        "outputId": "ac7fc447-9f61-4853-a173-dc7b0d737687",
        "collapsed": true
      },
      "id": "LuXji4Rgkh_Y",
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of male and female cases:\n",
            "    Male  Female\n",
            "0  43488   34542\n",
            "\n",
            "Age distribution:\n",
            "      Age  Count\n",
            "0    <NA>      0\n",
            "1       0   1625\n",
            "2       1   1019\n",
            "3       2   1305\n",
            "4       3   1177\n",
            "..    ...    ...\n",
            "96     95      9\n",
            "97     96      7\n",
            "98     97      3\n",
            "99     98      3\n",
            "100   100      1\n",
            "\n",
            "[101 rows x 2 columns]\n",
            "\n",
            "Age range:\n",
            "   min_age  max_age    avg_age\n",
            "0        0      100  43.435255\n",
            "\n",
            "Types and counts of fusions:\n",
            "                 Gene  Count\n",
            "0           BCR::ABL1    422\n",
            "1      RUNX1::RUNX1T1    134\n",
            "2           PML::RARA     96\n",
            "3         ETV6::RUNX1     84\n",
            "4         ETV6::NTRK3     74\n",
            "...               ...    ...\n",
            "34411   FOXP1::HOXA10      1\n",
            "34412     ZNF106::PML      1\n",
            "34413    CDK12::KCNRG      1\n",
            "34414    ATAD5::CDK12      1\n",
            "34415     PRDM16::SKI      1\n",
            "\n",
            "[34416 rows x 2 columns]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7d738ea0",
      "metadata": {
        "id": "7d738ea0"
      },
      "source": [
        "## Conclusion\n",
        "\n",
        "In this notebook we introduced basic SQL as a method for obtaining data in the Mitelman database BigQuery tables. We also provided detailed descriptions of the database tables to help with database navigation. In follow-up notebooks we will demonstrate table joins as a method for performing more complex queries."
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.9.6"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}